// lightningdevkit/structs/APIError.mjs
import { ShutdownScript } from '../structs/ShutdownScript.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * Indicates an error on the client's part (usually some variant of attempting to use too-low or
 * too-high values)
 */
export class APIError extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.APIError_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKAPIError_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new APIError_APIMisuseError(ptr);
            case 1: return new APIError_FeeRateTooHigh(ptr);
            case 2: return new APIError_InvalidRoute(ptr);
            case 3: return new APIError_ChannelUnavailable(ptr);
            case 4: return new APIError_MonitorUpdateInProgress(ptr);
            case 5: return new APIError_IncompatibleShutdownScript(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    clone_ptr() {
        const ret = bindings.APIError_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the APIError
     */
    clone() {
        const ret = bindings.APIError_clone(this.ptr);
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new APIMisuseError-variant APIError
     */
    static constructor_apimisuse_error(err) {
        const ret = bindings.APIError_apimisuse_error(bindings.encodeString(err));
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new FeeRateTooHigh-variant APIError
     */
    static constructor_fee_rate_too_high(err, feerate) {
        const ret = bindings.APIError_fee_rate_too_high(bindings.encodeString(err), feerate);
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidRoute-variant APIError
     */
    static constructor_invalid_route(err) {
        const ret = bindings.APIError_invalid_route(bindings.encodeString(err));
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new ChannelUnavailable-variant APIError
     */
    static constructor_channel_unavailable(err) {
        const ret = bindings.APIError_channel_unavailable(bindings.encodeString(err));
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new MonitorUpdateInProgress-variant APIError
     */
    static constructor_monitor_update_in_progress() {
        const ret = bindings.APIError_monitor_update_in_progress();
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new IncompatibleShutdownScript-variant APIError
     */
    static constructor_incompatible_shutdown_script(script) {
        const ret = bindings.APIError_incompatible_shutdown_script(CommonBase.get_ptr_of(script));
        const ret_hu_conv = APIError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Checks if two APIErrors contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     */
    eq(b) {
        const ret = bindings.APIError_eq(this.ptr, CommonBase.get_ptr_of(b));
        return ret;
    }
    /**
     * Serialize the APIError object into a byte array which can be read by APIError_read
     */
    write() {
        const ret = bindings.APIError_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
}
/** A APIError of type APIMisuseError */
export class APIError_APIMisuseError extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const err = bindings.LDKAPIError_APIMisuseError_get_err(ptr);
        const err_conv = bindings.decodeString(err);
        this.err = err_conv;
    }
}
/** A APIError of type FeeRateTooHigh */
export class APIError_FeeRateTooHigh extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const err = bindings.LDKAPIError_FeeRateTooHigh_get_err(ptr);
        const err_conv = bindings.decodeString(err);
        this.err = err_conv;
        this.feerate = bindings.LDKAPIError_FeeRateTooHigh_get_feerate(ptr);
    }
}
/** A APIError of type InvalidRoute */
export class APIError_InvalidRoute extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const err = bindings.LDKAPIError_InvalidRoute_get_err(ptr);
        const err_conv = bindings.decodeString(err);
        this.err = err_conv;
    }
}
/** A APIError of type ChannelUnavailable */
export class APIError_ChannelUnavailable extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const err = bindings.LDKAPIError_ChannelUnavailable_get_err(ptr);
        const err_conv = bindings.decodeString(err);
        this.err = err_conv;
    }
}
/** A APIError of type MonitorUpdateInProgress */
export class APIError_MonitorUpdateInProgress extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A APIError of type IncompatibleShutdownScript */
export class APIError_IncompatibleShutdownScript extends APIError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const script = bindings.LDKAPIError_IncompatibleShutdownScript_get_script(ptr);
        const script_hu_conv = new ShutdownScript(null, script);
        CommonBase.add_ref_from(script_hu_conv, this);
        this.script = script_hu_conv;
    }
}
