// lightningdevkit/structs/Bolt11InvoiceSignature.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * Recoverable signature
 */
export class Bolt11InvoiceSignature extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.Bolt11InvoiceSignature_free);
    }
    get_a() {
        const ret = bindings.Bolt11InvoiceSignature_get_a(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    set_a(val) {
        bindings.Bolt11InvoiceSignature_set_a(this.ptr, bindings.encodeUint8Array(bindings.check_arr_len(val, 68)));
    }
    /**
     * Constructs a new Bolt11InvoiceSignature given each field
     */
    static constructor_new(a_arg) {
        const ret = bindings.Bolt11InvoiceSignature_new(bindings.encodeUint8Array(bindings.check_arr_len(a_arg, 68)));
        const ret_hu_conv = new Bolt11InvoiceSignature(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.Bolt11InvoiceSignature_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the Bolt11InvoiceSignature
     */
    clone() {
        const ret = bindings.Bolt11InvoiceSignature_clone(this.ptr);
        const ret_hu_conv = new Bolt11InvoiceSignature(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the Bolt11InvoiceSignature.
     */
    hash() {
        const ret = bindings.Bolt11InvoiceSignature_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two Bolt11InvoiceSignatures contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.Bolt11InvoiceSignature_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
}
