// lightningdevkit/structs/Bolt11ParseError.mjs
import { Bech32Error } from '../structs/Bech32Error.mjs';
import { CommonBase, UnqualifiedError } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * Errors that indicate what is wrong with the invoice. They have some granularity for debug
 * reasons, but should generally result in an \"invalid BOLT11 invoice\" message for the user.
 */
export class Bolt11ParseError extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.Bolt11ParseError_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKBolt11ParseError_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Bolt11ParseError_Bech32Error(ptr);
            case 1: return new Bolt11ParseError_ParseAmountError(ptr);
            case 2: return new Bolt11ParseError_MalformedSignature(ptr);
            case 3: return new Bolt11ParseError_BadPrefix(ptr);
            case 4: return new Bolt11ParseError_UnknownCurrency(ptr);
            case 5: return new Bolt11ParseError_UnknownSiPrefix(ptr);
            case 6: return new Bolt11ParseError_MalformedHRP(ptr);
            case 7: return new Bolt11ParseError_TooShortDataPart(ptr);
            case 8: return new Bolt11ParseError_UnexpectedEndOfTaggedFields(ptr);
            case 9: return new Bolt11ParseError_DescriptionDecodeError(ptr);
            case 10: return new Bolt11ParseError_PaddingError(ptr);
            case 11: return new Bolt11ParseError_IntegerOverflowError(ptr);
            case 12: return new Bolt11ParseError_InvalidSegWitProgramLength(ptr);
            case 13: return new Bolt11ParseError_InvalidPubKeyHashLength(ptr);
            case 14: return new Bolt11ParseError_InvalidScriptHashLength(ptr);
            case 15: return new Bolt11ParseError_InvalidRecoveryId(ptr);
            case 16: return new Bolt11ParseError_InvalidSliceLength(ptr);
            case 17: return new Bolt11ParseError_Skip(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    clone_ptr() {
        const ret = bindings.Bolt11ParseError_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the Bolt11ParseError
     */
    clone() {
        const ret = bindings.Bolt11ParseError_clone(this.ptr);
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new Bech32Error-variant Bolt11ParseError
     */
    static constructor_bech32_error(a) {
        const ret = bindings.Bolt11ParseError_bech32_error(CommonBase.get_ptr_of(a));
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new ParseAmountError-variant Bolt11ParseError
     */
    static constructor_parse_amount_error(a) {
        const ret = bindings.Bolt11ParseError_parse_amount_error(0);
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new MalformedSignature-variant Bolt11ParseError
     */
    static constructor_malformed_signature(a) {
        const ret = bindings.Bolt11ParseError_malformed_signature(a);
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new BadPrefix-variant Bolt11ParseError
     */
    static constructor_bad_prefix() {
        const ret = bindings.Bolt11ParseError_bad_prefix();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new UnknownCurrency-variant Bolt11ParseError
     */
    static constructor_unknown_currency() {
        const ret = bindings.Bolt11ParseError_unknown_currency();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new UnknownSiPrefix-variant Bolt11ParseError
     */
    static constructor_unknown_si_prefix() {
        const ret = bindings.Bolt11ParseError_unknown_si_prefix();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new MalformedHRP-variant Bolt11ParseError
     */
    static constructor_malformed_hrp() {
        const ret = bindings.Bolt11ParseError_malformed_hrp();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new TooShortDataPart-variant Bolt11ParseError
     */
    static constructor_too_short_data_part() {
        const ret = bindings.Bolt11ParseError_too_short_data_part();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new UnexpectedEndOfTaggedFields-variant Bolt11ParseError
     */
    static constructor_unexpected_end_of_tagged_fields() {
        const ret = bindings.Bolt11ParseError_unexpected_end_of_tagged_fields();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new DescriptionDecodeError-variant Bolt11ParseError
     */
    static constructor_description_decode_error(a) {
        const ret = bindings.Bolt11ParseError_description_decode_error(0);
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new PaddingError-variant Bolt11ParseError
     */
    static constructor_padding_error() {
        const ret = bindings.Bolt11ParseError_padding_error();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new IntegerOverflowError-variant Bolt11ParseError
     */
    static constructor_integer_overflow_error() {
        const ret = bindings.Bolt11ParseError_integer_overflow_error();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidSegWitProgramLength-variant Bolt11ParseError
     */
    static constructor_invalid_seg_wit_program_length() {
        const ret = bindings.Bolt11ParseError_invalid_seg_wit_program_length();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidPubKeyHashLength-variant Bolt11ParseError
     */
    static constructor_invalid_pub_key_hash_length() {
        const ret = bindings.Bolt11ParseError_invalid_pub_key_hash_length();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidScriptHashLength-variant Bolt11ParseError
     */
    static constructor_invalid_script_hash_length() {
        const ret = bindings.Bolt11ParseError_invalid_script_hash_length();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidRecoveryId-variant Bolt11ParseError
     */
    static constructor_invalid_recovery_id() {
        const ret = bindings.Bolt11ParseError_invalid_recovery_id();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new InvalidSliceLength-variant Bolt11ParseError
     */
    static constructor_invalid_slice_length(a) {
        const ret = bindings.Bolt11ParseError_invalid_slice_length(bindings.encodeString(a));
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new Skip-variant Bolt11ParseError
     */
    static constructor_skip() {
        const ret = bindings.Bolt11ParseError_skip();
        const ret_hu_conv = Bolt11ParseError.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Checks if two Bolt11ParseErrors contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     */
    eq(b) {
        const ret = bindings.Bolt11ParseError_eq(this.ptr, CommonBase.get_ptr_of(b));
        return ret;
    }
    /**
     * Get the string representation of a Bolt11ParseError object
     */
    to_str() {
        const ret = bindings.Bolt11ParseError_to_str(this.ptr);
        const ret_conv = bindings.decodeString(ret);
        return ret_conv;
    }
}
/** A Bolt11ParseError of type Bech32Error */
export class Bolt11ParseError_Bech32Error extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const bech32_error = bindings.LDKBolt11ParseError_Bech32Error_get_bech32_error(ptr);
        const bech32_error_hu_conv = Bech32Error.constr_from_ptr(bech32_error);
        CommonBase.add_ref_from(bech32_error_hu_conv, this);
        this.bech32_error = bech32_error_hu_conv;
    }
}
/** A Bolt11ParseError of type ParseAmountError */
export class Bolt11ParseError_ParseAmountError extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const parse_amount_error = bindings.LDKBolt11ParseError_ParseAmountError_get_parse_amount_error(ptr);
        const parse_amount_error_conv = new UnqualifiedError(parse_amount_error);
        this.parse_amount_error = parse_amount_error_conv;
    }
}
/** A Bolt11ParseError of type MalformedSignature */
export class Bolt11ParseError_MalformedSignature extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        this.malformed_signature = bindings.LDKBolt11ParseError_MalformedSignature_get_malformed_signature(ptr);
    }
}
/** A Bolt11ParseError of type BadPrefix */
export class Bolt11ParseError_BadPrefix extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type UnknownCurrency */
export class Bolt11ParseError_UnknownCurrency extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type UnknownSiPrefix */
export class Bolt11ParseError_UnknownSiPrefix extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type MalformedHRP */
export class Bolt11ParseError_MalformedHRP extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type TooShortDataPart */
export class Bolt11ParseError_TooShortDataPart extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type UnexpectedEndOfTaggedFields */
export class Bolt11ParseError_UnexpectedEndOfTaggedFields extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type DescriptionDecodeError */
export class Bolt11ParseError_DescriptionDecodeError extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const description_decode_error = bindings.LDKBolt11ParseError_DescriptionDecodeError_get_description_decode_error(ptr);
        const description_decode_error_conv = new UnqualifiedError(description_decode_error);
        this.description_decode_error = description_decode_error_conv;
    }
}
/** A Bolt11ParseError of type PaddingError */
export class Bolt11ParseError_PaddingError extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type IntegerOverflowError */
export class Bolt11ParseError_IntegerOverflowError extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type InvalidSegWitProgramLength */
export class Bolt11ParseError_InvalidSegWitProgramLength extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type InvalidPubKeyHashLength */
export class Bolt11ParseError_InvalidPubKeyHashLength extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type InvalidScriptHashLength */
export class Bolt11ParseError_InvalidScriptHashLength extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type InvalidRecoveryId */
export class Bolt11ParseError_InvalidRecoveryId extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
/** A Bolt11ParseError of type InvalidSliceLength */
export class Bolt11ParseError_InvalidSliceLength extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const invalid_slice_length = bindings.LDKBolt11ParseError_InvalidSliceLength_get_invalid_slice_length(ptr);
        const invalid_slice_length_conv = bindings.decodeString(invalid_slice_length);
        this.invalid_slice_length = invalid_slice_length_conv;
    }
}
/** A Bolt11ParseError of type Skip */
export class Bolt11ParseError_Skip extends Bolt11ParseError {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
