// lightningdevkit/structs/ChannelTypeFeatures.mjs
import { Result_NoneNoneZ } from '../structs/Result_NoneNoneZ.mjs';
import { Result_ChannelTypeFeaturesDecodeErrorZ } from '../structs/Result_ChannelTypeFeaturesDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * Features used within the channel_type field in an OpenChannel message.
 *
 * A channel is always of some known \"type\", describing the transaction formats used and the exact
 * semantics of our interaction with our peer.
 *
 * Note that because a channel is a specific type which is proposed by the opener and accepted by
 * the counterparty, only required features are allowed here.
 *
 * This is serialized differently from other feature types - it is not prefixed by a length, and
 * thus must only appear inside a TLV where its length is known in advance.
 */
export class ChannelTypeFeatures extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.ChannelTypeFeatures_free);
    }
    /**
     * Serialize the ChannelTypeFeatures object into a byte array which can be read by ChannelTypeFeatures_read
     */
    write() {
        const ret = bindings.ChannelTypeFeatures_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a ChannelTypeFeatures from a byte array, created by ChannelTypeFeatures_write
     */
    static constructor_read(ser) {
        const ret = bindings.ChannelTypeFeatures_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_ChannelTypeFeaturesDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if two ChannelTypeFeaturess contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.ChannelTypeFeatures_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.ChannelTypeFeatures_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the ChannelTypeFeatures
     */
    clone() {
        const ret = bindings.ChannelTypeFeatures_clone(this.ptr);
        const ret_hu_conv = new ChannelTypeFeatures(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the ChannelTypeFeatures.
     */
    hash() {
        const ret = bindings.ChannelTypeFeatures_hash(this.ptr);
        return ret;
    }
    /**
     * Constructs a ChannelTypeFeatures with only static_remotekey set
     */
    static constructor_only_static_remote_key() {
        const ret = bindings.ChannelTypeFeatures_only_static_remote_key();
        const ret_hu_conv = new ChannelTypeFeatures(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a ChannelTypeFeatures with anchors support
     */
    static constructor_anchors_zero_htlc_fee_and_dependencies() {
        const ret = bindings.ChannelTypeFeatures_anchors_zero_htlc_fee_and_dependencies();
        const ret_hu_conv = new ChannelTypeFeatures(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Create a blank Features with no features set
     */
    static constructor_empty() {
        const ret = bindings.ChannelTypeFeatures_empty();
        const ret_hu_conv = new ChannelTypeFeatures(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Returns the feature set as a list of bytes, in little-endian. This is in reverse byte order
     * from most on-the-wire encodings.
     */
    le_flags() {
        const ret = bindings.ChannelTypeFeatures_le_flags(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Returns true if this `Features` has any optional flags set
     */
    supports_any_optional_bits() {
        const ret = bindings.ChannelTypeFeatures_supports_any_optional_bits(this.ptr);
        return ret;
    }
    /**
     * Returns true if this `Features` object contains required features unknown by `other`.
     */
    requires_unknown_bits_from(other) {
        const ret = bindings.ChannelTypeFeatures_requires_unknown_bits_from(this.ptr, CommonBase.get_ptr_of(other));
        CommonBase.add_ref_from(this, other);
        return ret;
    }
    /**
     * Returns the set of required features unknown by `other`, as their bit position.
     */
    required_unknown_bits_from(other) {
        const ret = bindings.ChannelTypeFeatures_required_unknown_bits_from(this.ptr, CommonBase.get_ptr_of(other));
        const ret_conv = bindings.decodeUint64Array(ret);
        CommonBase.add_ref_from(this, other);
        return ret_conv;
    }
    /**
     * Returns true if this `Features` object contains unknown feature flags which are set as
     * \"required\".
     */
    requires_unknown_bits() {
        const ret = bindings.ChannelTypeFeatures_requires_unknown_bits(this.ptr);
        return ret;
    }
    /**
     * Returns true if this `Features` supports any bits which we do not know of
     */
    supports_unknown_bits() {
        const ret = bindings.ChannelTypeFeatures_supports_unknown_bits(this.ptr);
        return ret;
    }
    /**
     * Sets a required feature bit. Errors if `bit` is outside the feature range as defined
     * by [BOLT 9].
     *
     * Note: Required bits are even. If an odd bit is given, then the corresponding even bit will
     * be set instead (i.e., `bit - 1`).
     *
     * [BOLT 9]: https://github.com/lightning/bolts/blob/master/09-features.md
     */
    set_required_feature_bit(bit) {
        const ret = bindings.ChannelTypeFeatures_set_required_feature_bit(this.ptr, bit);
        const ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Sets an optional feature bit. Errors if `bit` is outside the feature range as defined
     * by [BOLT 9].
     *
     * Note: Optional bits are odd. If an even bit is given, then the corresponding odd bit will be
     * set instead (i.e., `bit + 1`).
     *
     * [BOLT 9]: https://github.com/lightning/bolts/blob/master/09-features.md
     */
    set_optional_feature_bit(bit) {
        const ret = bindings.ChannelTypeFeatures_set_optional_feature_bit(this.ptr, bit);
        const ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Sets a required custom feature bit. Errors if `bit` is outside the custom range as defined
     * by [bLIP 2] or if it is a known `T` feature.
     *
     * Note: Required bits are even. If an odd bit is given, then the corresponding even bit will
     * be set instead (i.e., `bit - 1`).
     *
     * [bLIP 2]: https://github.com/lightning/blips/blob/master/blip-0002.md#feature-bits
     */
    set_required_custom_bit(bit) {
        const ret = bindings.ChannelTypeFeatures_set_required_custom_bit(this.ptr, bit);
        const ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Sets an optional custom feature bit. Errors if `bit` is outside the custom range as defined
     * by [bLIP 2] or if it is a known `T` feature.
     *
     * Note: Optional bits are odd. If an even bit is given, then the corresponding odd bit will be
     * set instead (i.e., `bit + 1`).
     *
     * [bLIP 2]: https://github.com/lightning/blips/blob/master/blip-0002.md#feature-bits
     */
    set_optional_custom_bit(bit) {
        const ret = bindings.ChannelTypeFeatures_set_optional_custom_bit(this.ptr, bit);
        const ret_hu_conv = Result_NoneNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Unsets the `scid_privacy` feature
     */
    clear_scid_privacy() {
        bindings.ChannelTypeFeatures_clear_scid_privacy(this.ptr);
    }
    /**
     * Unsets the `anchors_zero_fee_htlc_tx` feature
     */
    clear_anchors_zero_fee_htlc_tx() {
        bindings.ChannelTypeFeatures_clear_anchors_zero_fee_htlc_tx(this.ptr);
    }
    /**
     * Set this feature as optional.
     */
    set_static_remote_key_optional() {
        bindings.ChannelTypeFeatures_set_static_remote_key_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_static_remote_key_required() {
        bindings.ChannelTypeFeatures_set_static_remote_key_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_static_remote_key() {
        const ret = bindings.ChannelTypeFeatures_supports_static_remote_key(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_static_remote_key() {
        const ret = bindings.ChannelTypeFeatures_requires_static_remote_key(this.ptr);
        return ret;
    }
    /**
     * Set this feature as optional.
     */
    set_anchors_nonzero_fee_htlc_tx_optional() {
        bindings.ChannelTypeFeatures_set_anchors_nonzero_fee_htlc_tx_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_anchors_nonzero_fee_htlc_tx_required() {
        bindings.ChannelTypeFeatures_set_anchors_nonzero_fee_htlc_tx_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_anchors_nonzero_fee_htlc_tx() {
        const ret = bindings.ChannelTypeFeatures_supports_anchors_nonzero_fee_htlc_tx(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_anchors_nonzero_fee_htlc_tx() {
        const ret = bindings.ChannelTypeFeatures_requires_anchors_nonzero_fee_htlc_tx(this.ptr);
        return ret;
    }
    /**
     * Set this feature as optional.
     */
    set_anchors_zero_fee_htlc_tx_optional() {
        bindings.ChannelTypeFeatures_set_anchors_zero_fee_htlc_tx_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_anchors_zero_fee_htlc_tx_required() {
        bindings.ChannelTypeFeatures_set_anchors_zero_fee_htlc_tx_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_anchors_zero_fee_htlc_tx() {
        const ret = bindings.ChannelTypeFeatures_supports_anchors_zero_fee_htlc_tx(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_anchors_zero_fee_htlc_tx() {
        const ret = bindings.ChannelTypeFeatures_requires_anchors_zero_fee_htlc_tx(this.ptr);
        return ret;
    }
    /**
     * Set this feature as optional.
     */
    set_taproot_optional() {
        bindings.ChannelTypeFeatures_set_taproot_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_taproot_required() {
        bindings.ChannelTypeFeatures_set_taproot_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_taproot() {
        const ret = bindings.ChannelTypeFeatures_supports_taproot(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_taproot() {
        const ret = bindings.ChannelTypeFeatures_requires_taproot(this.ptr);
        return ret;
    }
    /**
     * Set this feature as optional.
     */
    set_scid_privacy_optional() {
        bindings.ChannelTypeFeatures_set_scid_privacy_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_scid_privacy_required() {
        bindings.ChannelTypeFeatures_set_scid_privacy_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_scid_privacy() {
        const ret = bindings.ChannelTypeFeatures_supports_scid_privacy(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_scid_privacy() {
        const ret = bindings.ChannelTypeFeatures_requires_scid_privacy(this.ptr);
        return ret;
    }
    /**
     * Set this feature as optional.
     */
    set_zero_conf_optional() {
        bindings.ChannelTypeFeatures_set_zero_conf_optional(this.ptr);
    }
    /**
     * Set this feature as required.
     */
    set_zero_conf_required() {
        bindings.ChannelTypeFeatures_set_zero_conf_required(this.ptr);
    }
    /**
     * Checks if this feature is supported.
     */
    supports_zero_conf() {
        const ret = bindings.ChannelTypeFeatures_supports_zero_conf(this.ptr);
        return ret;
    }
    /**
     * Checks if this feature is required.
     */
    requires_zero_conf() {
        const ret = bindings.ChannelTypeFeatures_requires_zero_conf(this.ptr);
        return ret;
    }
}
