// lightningdevkit/structs/ChannelUpdate.mjs
import { UnsignedChannelUpdate } from '../structs/UnsignedChannelUpdate.mjs';
import { Result_ChannelUpdateDecodeErrorZ } from '../structs/Result_ChannelUpdateDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * A [`channel_update`] message to be sent to or received from a peer.
 *
 * [`channel_update`]: https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-channel_update-message
 */
export class ChannelUpdate extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.ChannelUpdate_free);
    }
    /**
     * A signature of the channel update
     */
    get_signature() {
        const ret = bindings.ChannelUpdate_get_signature(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * A signature of the channel update
     */
    set_signature(val) {
        bindings.ChannelUpdate_set_signature(this.ptr, bindings.encodeUint8Array(bindings.check_arr_len(val, 64)));
    }
    /**
     * The actual channel update
     */
    get_contents() {
        const ret = bindings.ChannelUpdate_get_contents(this.ptr);
        const ret_hu_conv = new UnsignedChannelUpdate(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * The actual channel update
     */
    set_contents(val) {
        bindings.ChannelUpdate_set_contents(this.ptr, CommonBase.get_ptr_of(val));
    }
    /**
     * Constructs a new ChannelUpdate given each field
     */
    static constructor_new(signature_arg, contents_arg) {
        const ret = bindings.ChannelUpdate_new(bindings.encodeUint8Array(bindings.check_arr_len(signature_arg, 64)), CommonBase.get_ptr_of(contents_arg));
        const ret_hu_conv = new ChannelUpdate(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.ChannelUpdate_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the ChannelUpdate
     */
    clone() {
        const ret = bindings.ChannelUpdate_clone(this.ptr);
        const ret_hu_conv = new ChannelUpdate(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the ChannelUpdate.
     */
    hash() {
        const ret = bindings.ChannelUpdate_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two ChannelUpdates contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.ChannelUpdate_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Serialize the ChannelUpdate object into a byte array which can be read by ChannelUpdate_read
     */
    write() {
        const ret = bindings.ChannelUpdate_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a ChannelUpdate from a byte array, created by ChannelUpdate_write
     */
    static constructor_read(ser) {
        const ret = bindings.ChannelUpdate_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_ChannelUpdateDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
