// lightningdevkit/structs/MessageSendEvent.mjs
import { ChannelAnnouncement } from '../structs/ChannelAnnouncement.mjs';
import { ChannelUpdate } from '../structs/ChannelUpdate.mjs';
import { AcceptChannel } from '../structs/AcceptChannel.mjs';
import { AcceptChannelV2 } from '../structs/AcceptChannelV2.mjs';
import { OpenChannel } from '../structs/OpenChannel.mjs';
import { OpenChannelV2 } from '../structs/OpenChannelV2.mjs';
import { FundingCreated } from '../structs/FundingCreated.mjs';
import { FundingSigned } from '../structs/FundingSigned.mjs';
import { Stfu } from '../structs/Stfu.mjs';
import { SpliceInit } from '../structs/SpliceInit.mjs';
import { SpliceAck } from '../structs/SpliceAck.mjs';
import { SpliceLocked } from '../structs/SpliceLocked.mjs';
import { TxAddInput } from '../structs/TxAddInput.mjs';
import { TxAddOutput } from '../structs/TxAddOutput.mjs';
import { TxRemoveInput } from '../structs/TxRemoveInput.mjs';
import { TxRemoveOutput } from '../structs/TxRemoveOutput.mjs';
import { TxComplete } from '../structs/TxComplete.mjs';
import { TxSignatures } from '../structs/TxSignatures.mjs';
import { TxInitRbf } from '../structs/TxInitRbf.mjs';
import { TxAckRbf } from '../structs/TxAckRbf.mjs';
import { TxAbort } from '../structs/TxAbort.mjs';
import { ChannelReady } from '../structs/ChannelReady.mjs';
import { AnnouncementSignatures } from '../structs/AnnouncementSignatures.mjs';
import { CommitmentUpdate } from '../structs/CommitmentUpdate.mjs';
import { RevokeAndACK } from '../structs/RevokeAndACK.mjs';
import { ClosingSigned } from '../structs/ClosingSigned.mjs';
import { Shutdown } from '../structs/Shutdown.mjs';
import { ChannelReestablish } from '../structs/ChannelReestablish.mjs';
import { NodeAnnouncement } from '../structs/NodeAnnouncement.mjs';
import { ErrorAction } from '../structs/ErrorAction.mjs';
import { QueryChannelRange } from '../structs/QueryChannelRange.mjs';
import { QueryShortChannelIds } from '../structs/QueryShortChannelIds.mjs';
import { ReplyChannelRange } from '../structs/ReplyChannelRange.mjs';
import { GossipTimestampFilter } from '../structs/GossipTimestampFilter.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An event generated by ChannelManager which indicates a message should be sent to a peer (or
 * broadcast to most peers).
 * These events are handled by PeerManager::process_events if you are using a PeerManager.
 */
export class MessageSendEvent extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.MessageSendEvent_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKMessageSendEvent_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new MessageSendEvent_SendAcceptChannel(ptr);
            case 1: return new MessageSendEvent_SendAcceptChannelV2(ptr);
            case 2: return new MessageSendEvent_SendOpenChannel(ptr);
            case 3: return new MessageSendEvent_SendOpenChannelV2(ptr);
            case 4: return new MessageSendEvent_SendFundingCreated(ptr);
            case 5: return new MessageSendEvent_SendFundingSigned(ptr);
            case 6: return new MessageSendEvent_SendStfu(ptr);
            case 7: return new MessageSendEvent_SendSpliceInit(ptr);
            case 8: return new MessageSendEvent_SendSpliceAck(ptr);
            case 9: return new MessageSendEvent_SendSpliceLocked(ptr);
            case 10: return new MessageSendEvent_SendTxAddInput(ptr);
            case 11: return new MessageSendEvent_SendTxAddOutput(ptr);
            case 12: return new MessageSendEvent_SendTxRemoveInput(ptr);
            case 13: return new MessageSendEvent_SendTxRemoveOutput(ptr);
            case 14: return new MessageSendEvent_SendTxComplete(ptr);
            case 15: return new MessageSendEvent_SendTxSignatures(ptr);
            case 16: return new MessageSendEvent_SendTxInitRbf(ptr);
            case 17: return new MessageSendEvent_SendTxAckRbf(ptr);
            case 18: return new MessageSendEvent_SendTxAbort(ptr);
            case 19: return new MessageSendEvent_SendChannelReady(ptr);
            case 20: return new MessageSendEvent_SendAnnouncementSignatures(ptr);
            case 21: return new MessageSendEvent_UpdateHTLCs(ptr);
            case 22: return new MessageSendEvent_SendRevokeAndACK(ptr);
            case 23: return new MessageSendEvent_SendClosingSigned(ptr);
            case 24: return new MessageSendEvent_SendShutdown(ptr);
            case 25: return new MessageSendEvent_SendChannelReestablish(ptr);
            case 26: return new MessageSendEvent_SendChannelAnnouncement(ptr);
            case 27: return new MessageSendEvent_BroadcastChannelAnnouncement(ptr);
            case 28: return new MessageSendEvent_BroadcastChannelUpdate(ptr);
            case 29: return new MessageSendEvent_BroadcastNodeAnnouncement(ptr);
            case 30: return new MessageSendEvent_SendChannelUpdate(ptr);
            case 31: return new MessageSendEvent_HandleError(ptr);
            case 32: return new MessageSendEvent_SendChannelRangeQuery(ptr);
            case 33: return new MessageSendEvent_SendShortIdsQuery(ptr);
            case 34: return new MessageSendEvent_SendReplyChannelRange(ptr);
            case 35: return new MessageSendEvent_SendGossipTimestampFilter(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    clone_ptr() {
        const ret = bindings.MessageSendEvent_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the MessageSendEvent
     */
    clone() {
        const ret = bindings.MessageSendEvent_clone(this.ptr);
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendAcceptChannel-variant MessageSendEvent
     */
    static constructor_send_accept_channel(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_accept_channel(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendAcceptChannelV2-variant MessageSendEvent
     */
    static constructor_send_accept_channel_v2(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_accept_channel_v2(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendOpenChannel-variant MessageSendEvent
     */
    static constructor_send_open_channel(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_open_channel(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendOpenChannelV2-variant MessageSendEvent
     */
    static constructor_send_open_channel_v2(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_open_channel_v2(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendFundingCreated-variant MessageSendEvent
     */
    static constructor_send_funding_created(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_funding_created(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendFundingSigned-variant MessageSendEvent
     */
    static constructor_send_funding_signed(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_funding_signed(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendStfu-variant MessageSendEvent
     */
    static constructor_send_stfu(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_stfu(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendSpliceInit-variant MessageSendEvent
     */
    static constructor_send_splice_init(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_splice_init(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendSpliceAck-variant MessageSendEvent
     */
    static constructor_send_splice_ack(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_splice_ack(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendSpliceLocked-variant MessageSendEvent
     */
    static constructor_send_splice_locked(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_splice_locked(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxAddInput-variant MessageSendEvent
     */
    static constructor_send_tx_add_input(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_add_input(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxAddOutput-variant MessageSendEvent
     */
    static constructor_send_tx_add_output(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_add_output(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxRemoveInput-variant MessageSendEvent
     */
    static constructor_send_tx_remove_input(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_remove_input(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxRemoveOutput-variant MessageSendEvent
     */
    static constructor_send_tx_remove_output(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_remove_output(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxComplete-variant MessageSendEvent
     */
    static constructor_send_tx_complete(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_complete(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxSignatures-variant MessageSendEvent
     */
    static constructor_send_tx_signatures(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_signatures(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxInitRbf-variant MessageSendEvent
     */
    static constructor_send_tx_init_rbf(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_init_rbf(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxAckRbf-variant MessageSendEvent
     */
    static constructor_send_tx_ack_rbf(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_ack_rbf(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendTxAbort-variant MessageSendEvent
     */
    static constructor_send_tx_abort(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_tx_abort(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendChannelReady-variant MessageSendEvent
     */
    static constructor_send_channel_ready(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_channel_ready(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendAnnouncementSignatures-variant MessageSendEvent
     */
    static constructor_send_announcement_signatures(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_announcement_signatures(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new UpdateHTLCs-variant MessageSendEvent
     */
    static constructor_update_htlcs(node_id, updates) {
        const ret = bindings.MessageSendEvent_update_htlcs(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(updates));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendRevokeAndACK-variant MessageSendEvent
     */
    static constructor_send_revoke_and_ack(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_revoke_and_ack(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendClosingSigned-variant MessageSendEvent
     */
    static constructor_send_closing_signed(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_closing_signed(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendShutdown-variant MessageSendEvent
     */
    static constructor_send_shutdown(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_shutdown(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendChannelReestablish-variant MessageSendEvent
     */
    static constructor_send_channel_reestablish(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_channel_reestablish(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendChannelAnnouncement-variant MessageSendEvent
     */
    static constructor_send_channel_announcement(node_id, msg, update_msg) {
        const ret = bindings.MessageSendEvent_send_channel_announcement(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg), CommonBase.get_ptr_of(update_msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new BroadcastChannelAnnouncement-variant MessageSendEvent
     */
    static constructor_broadcast_channel_announcement(msg, update_msg) {
        const ret = bindings.MessageSendEvent_broadcast_channel_announcement(CommonBase.get_ptr_of(msg), CommonBase.get_ptr_of(update_msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new BroadcastChannelUpdate-variant MessageSendEvent
     */
    static constructor_broadcast_channel_update(msg) {
        const ret = bindings.MessageSendEvent_broadcast_channel_update(CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new BroadcastNodeAnnouncement-variant MessageSendEvent
     */
    static constructor_broadcast_node_announcement(msg) {
        const ret = bindings.MessageSendEvent_broadcast_node_announcement(CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendChannelUpdate-variant MessageSendEvent
     */
    static constructor_send_channel_update(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_channel_update(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new HandleError-variant MessageSendEvent
     */
    static constructor_handle_error(node_id, action) {
        const ret = bindings.MessageSendEvent_handle_error(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(action));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendChannelRangeQuery-variant MessageSendEvent
     */
    static constructor_send_channel_range_query(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_channel_range_query(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendShortIdsQuery-variant MessageSendEvent
     */
    static constructor_send_short_ids_query(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_short_ids_query(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendReplyChannelRange-variant MessageSendEvent
     */
    static constructor_send_reply_channel_range(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_reply_channel_range(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new SendGossipTimestampFilter-variant MessageSendEvent
     */
    static constructor_send_gossip_timestamp_filter(node_id, msg) {
        const ret = bindings.MessageSendEvent_send_gossip_timestamp_filter(bindings.encodeUint8Array(bindings.check_arr_len(node_id, 33)), CommonBase.get_ptr_of(msg));
        const ret_hu_conv = MessageSendEvent.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
}
/** A MessageSendEvent of type SendAcceptChannel */
export class MessageSendEvent_SendAcceptChannel extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendAcceptChannel_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendAcceptChannel_get_msg(ptr);
        const msg_hu_conv = new AcceptChannel(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendAcceptChannelV2 */
export class MessageSendEvent_SendAcceptChannelV2 extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendAcceptChannelV2_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendAcceptChannelV2_get_msg(ptr);
        const msg_hu_conv = new AcceptChannelV2(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendOpenChannel */
export class MessageSendEvent_SendOpenChannel extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendOpenChannel_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendOpenChannel_get_msg(ptr);
        const msg_hu_conv = new OpenChannel(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendOpenChannelV2 */
export class MessageSendEvent_SendOpenChannelV2 extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendOpenChannelV2_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendOpenChannelV2_get_msg(ptr);
        const msg_hu_conv = new OpenChannelV2(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendFundingCreated */
export class MessageSendEvent_SendFundingCreated extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendFundingCreated_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendFundingCreated_get_msg(ptr);
        const msg_hu_conv = new FundingCreated(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendFundingSigned */
export class MessageSendEvent_SendFundingSigned extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendFundingSigned_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendFundingSigned_get_msg(ptr);
        const msg_hu_conv = new FundingSigned(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendStfu */
export class MessageSendEvent_SendStfu extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendStfu_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendStfu_get_msg(ptr);
        const msg_hu_conv = new Stfu(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendSpliceInit */
export class MessageSendEvent_SendSpliceInit extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendSpliceInit_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendSpliceInit_get_msg(ptr);
        const msg_hu_conv = new SpliceInit(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendSpliceAck */
export class MessageSendEvent_SendSpliceAck extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendSpliceAck_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendSpliceAck_get_msg(ptr);
        const msg_hu_conv = new SpliceAck(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendSpliceLocked */
export class MessageSendEvent_SendSpliceLocked extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendSpliceLocked_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendSpliceLocked_get_msg(ptr);
        const msg_hu_conv = new SpliceLocked(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxAddInput */
export class MessageSendEvent_SendTxAddInput extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxAddInput_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxAddInput_get_msg(ptr);
        const msg_hu_conv = new TxAddInput(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxAddOutput */
export class MessageSendEvent_SendTxAddOutput extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxAddOutput_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxAddOutput_get_msg(ptr);
        const msg_hu_conv = new TxAddOutput(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxRemoveInput */
export class MessageSendEvent_SendTxRemoveInput extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxRemoveInput_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxRemoveInput_get_msg(ptr);
        const msg_hu_conv = new TxRemoveInput(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxRemoveOutput */
export class MessageSendEvent_SendTxRemoveOutput extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxRemoveOutput_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxRemoveOutput_get_msg(ptr);
        const msg_hu_conv = new TxRemoveOutput(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxComplete */
export class MessageSendEvent_SendTxComplete extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxComplete_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxComplete_get_msg(ptr);
        const msg_hu_conv = new TxComplete(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxSignatures */
export class MessageSendEvent_SendTxSignatures extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxSignatures_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxSignatures_get_msg(ptr);
        const msg_hu_conv = new TxSignatures(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxInitRbf */
export class MessageSendEvent_SendTxInitRbf extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxInitRbf_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxInitRbf_get_msg(ptr);
        const msg_hu_conv = new TxInitRbf(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxAckRbf */
export class MessageSendEvent_SendTxAckRbf extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxAckRbf_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxAckRbf_get_msg(ptr);
        const msg_hu_conv = new TxAckRbf(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendTxAbort */
export class MessageSendEvent_SendTxAbort extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendTxAbort_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendTxAbort_get_msg(ptr);
        const msg_hu_conv = new TxAbort(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendChannelReady */
export class MessageSendEvent_SendChannelReady extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendChannelReady_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendChannelReady_get_msg(ptr);
        const msg_hu_conv = new ChannelReady(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendAnnouncementSignatures */
export class MessageSendEvent_SendAnnouncementSignatures extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendAnnouncementSignatures_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendAnnouncementSignatures_get_msg(ptr);
        const msg_hu_conv = new AnnouncementSignatures(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type UpdateHTLCs */
export class MessageSendEvent_UpdateHTLCs extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_UpdateHTLCs_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const updates = bindings.LDKMessageSendEvent_UpdateHTLCs_get_updates(ptr);
        const updates_hu_conv = new CommitmentUpdate(null, updates);
        CommonBase.add_ref_from(updates_hu_conv, this);
        this.updates = updates_hu_conv;
    }
}
/** A MessageSendEvent of type SendRevokeAndACK */
export class MessageSendEvent_SendRevokeAndACK extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendRevokeAndACK_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendRevokeAndACK_get_msg(ptr);
        const msg_hu_conv = new RevokeAndACK(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendClosingSigned */
export class MessageSendEvent_SendClosingSigned extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendClosingSigned_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendClosingSigned_get_msg(ptr);
        const msg_hu_conv = new ClosingSigned(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendShutdown */
export class MessageSendEvent_SendShutdown extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendShutdown_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendShutdown_get_msg(ptr);
        const msg_hu_conv = new Shutdown(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendChannelReestablish */
export class MessageSendEvent_SendChannelReestablish extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendChannelReestablish_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendChannelReestablish_get_msg(ptr);
        const msg_hu_conv = new ChannelReestablish(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendChannelAnnouncement */
export class MessageSendEvent_SendChannelAnnouncement extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendChannelAnnouncement_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendChannelAnnouncement_get_msg(ptr);
        const msg_hu_conv = new ChannelAnnouncement(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
        const update_msg = bindings.LDKMessageSendEvent_SendChannelAnnouncement_get_update_msg(ptr);
        const update_msg_hu_conv = new ChannelUpdate(null, update_msg);
        CommonBase.add_ref_from(update_msg_hu_conv, this);
        this.update_msg = update_msg_hu_conv;
    }
}
/** A MessageSendEvent of type BroadcastChannelAnnouncement */
export class MessageSendEvent_BroadcastChannelAnnouncement extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const msg = bindings.LDKMessageSendEvent_BroadcastChannelAnnouncement_get_msg(ptr);
        const msg_hu_conv = new ChannelAnnouncement(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
        const update_msg = bindings.LDKMessageSendEvent_BroadcastChannelAnnouncement_get_update_msg(ptr);
        const update_msg_hu_conv = new ChannelUpdate(null, update_msg);
        CommonBase.add_ref_from(update_msg_hu_conv, this);
        this.update_msg = update_msg_hu_conv;
    }
}
/** A MessageSendEvent of type BroadcastChannelUpdate */
export class MessageSendEvent_BroadcastChannelUpdate extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const msg = bindings.LDKMessageSendEvent_BroadcastChannelUpdate_get_msg(ptr);
        const msg_hu_conv = new ChannelUpdate(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type BroadcastNodeAnnouncement */
export class MessageSendEvent_BroadcastNodeAnnouncement extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const msg = bindings.LDKMessageSendEvent_BroadcastNodeAnnouncement_get_msg(ptr);
        const msg_hu_conv = new NodeAnnouncement(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendChannelUpdate */
export class MessageSendEvent_SendChannelUpdate extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendChannelUpdate_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendChannelUpdate_get_msg(ptr);
        const msg_hu_conv = new ChannelUpdate(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type HandleError */
export class MessageSendEvent_HandleError extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_HandleError_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const action = bindings.LDKMessageSendEvent_HandleError_get_action(ptr);
        const action_hu_conv = ErrorAction.constr_from_ptr(action);
        CommonBase.add_ref_from(action_hu_conv, this);
        this.action = action_hu_conv;
    }
}
/** A MessageSendEvent of type SendChannelRangeQuery */
export class MessageSendEvent_SendChannelRangeQuery extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendChannelRangeQuery_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendChannelRangeQuery_get_msg(ptr);
        const msg_hu_conv = new QueryChannelRange(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendShortIdsQuery */
export class MessageSendEvent_SendShortIdsQuery extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendShortIdsQuery_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendShortIdsQuery_get_msg(ptr);
        const msg_hu_conv = new QueryShortChannelIds(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendReplyChannelRange */
export class MessageSendEvent_SendReplyChannelRange extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendReplyChannelRange_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendReplyChannelRange_get_msg(ptr);
        const msg_hu_conv = new ReplyChannelRange(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
/** A MessageSendEvent of type SendGossipTimestampFilter */
export class MessageSendEvent_SendGossipTimestampFilter extends MessageSendEvent {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const node_id = bindings.LDKMessageSendEvent_SendGossipTimestampFilter_get_node_id(ptr);
        const node_id_conv = bindings.decodeUint8Array(node_id);
        this.node_id = node_id_conv;
        const msg = bindings.LDKMessageSendEvent_SendGossipTimestampFilter_get_msg(ptr);
        const msg_hu_conv = new GossipTimestampFilter(null, msg);
        CommonBase.add_ref_from(msg_hu_conv, this);
        this.msg = msg_hu_conv;
    }
}
