// lightningdevkit/structs/MinFinalCltvExpiryDelta.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * `min_final_cltv_expiry_delta` to use for the last HTLC in the route
 */
export class MinFinalCltvExpiryDelta extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.MinFinalCltvExpiryDelta_free);
    }
    get_a() {
        const ret = bindings.MinFinalCltvExpiryDelta_get_a(this.ptr);
        return ret;
    }
    set_a(val) {
        bindings.MinFinalCltvExpiryDelta_set_a(this.ptr, val);
    }
    /**
     * Constructs a new MinFinalCltvExpiryDelta given each field
     */
    static constructor_new(a_arg) {
        const ret = bindings.MinFinalCltvExpiryDelta_new(a_arg);
        const ret_hu_conv = new MinFinalCltvExpiryDelta(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.MinFinalCltvExpiryDelta_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the MinFinalCltvExpiryDelta
     */
    clone() {
        const ret = bindings.MinFinalCltvExpiryDelta_clone(this.ptr);
        const ret_hu_conv = new MinFinalCltvExpiryDelta(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the MinFinalCltvExpiryDelta.
     */
    hash() {
        const ret = bindings.MinFinalCltvExpiryDelta_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two MinFinalCltvExpiryDeltas contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.MinFinalCltvExpiryDelta_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
}
