// lightningdevkit/structs/NodeAnnouncement.mjs
import { UnsignedNodeAnnouncement } from '../structs/UnsignedNodeAnnouncement.mjs';
import { Result_NodeAnnouncementDecodeErrorZ } from '../structs/Result_NodeAnnouncementDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * A [`node_announcement`] message to be sent to or received from a peer.
 *
 * [`node_announcement`]: https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#the-node_announcement-message
 */
export class NodeAnnouncement extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.NodeAnnouncement_free);
    }
    /**
     * The signature by the node key
     */
    get_signature() {
        const ret = bindings.NodeAnnouncement_get_signature(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * The signature by the node key
     */
    set_signature(val) {
        bindings.NodeAnnouncement_set_signature(this.ptr, bindings.encodeUint8Array(bindings.check_arr_len(val, 64)));
    }
    /**
     * The actual content of the announcement
     */
    get_contents() {
        const ret = bindings.NodeAnnouncement_get_contents(this.ptr);
        const ret_hu_conv = new UnsignedNodeAnnouncement(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * The actual content of the announcement
     */
    set_contents(val) {
        bindings.NodeAnnouncement_set_contents(this.ptr, CommonBase.get_ptr_of(val));
    }
    /**
     * Constructs a new NodeAnnouncement given each field
     */
    static constructor_new(signature_arg, contents_arg) {
        const ret = bindings.NodeAnnouncement_new(bindings.encodeUint8Array(bindings.check_arr_len(signature_arg, 64)), CommonBase.get_ptr_of(contents_arg));
        const ret_hu_conv = new NodeAnnouncement(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.NodeAnnouncement_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the NodeAnnouncement
     */
    clone() {
        const ret = bindings.NodeAnnouncement_clone(this.ptr);
        const ret_hu_conv = new NodeAnnouncement(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the NodeAnnouncement.
     */
    hash() {
        const ret = bindings.NodeAnnouncement_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two NodeAnnouncements contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.NodeAnnouncement_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Serialize the NodeAnnouncement object into a byte array which can be read by NodeAnnouncement_read
     */
    write() {
        const ret = bindings.NodeAnnouncement_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a NodeAnnouncement from a byte array, created by NodeAnnouncement_write
     */
    static constructor_read(ser) {
        const ret = bindings.NodeAnnouncement_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_NodeAnnouncementDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
