// lightningdevkit/structs/NodeId.mjs
import { Result_NodeIdDecodeErrorZ } from '../structs/Result_NodeIdDecodeErrorZ.mjs';
import { Result_PublicKeySecp256k1ErrorZ } from '../structs/Result_PublicKeySecp256k1ErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * Represents the compressed public key of a node
 */
export class NodeId extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.NodeId_free);
    }
    clone_ptr() {
        const ret = bindings.NodeId_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the NodeId
     */
    clone() {
        const ret = bindings.NodeId_clone(this.ptr);
        const ret_hu_conv = new NodeId(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Checks if two NodeIds contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.NodeId_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Create a new NodeId from a public key
     */
    static constructor_from_pubkey(pubkey) {
        const ret = bindings.NodeId_from_pubkey(bindings.encodeUint8Array(bindings.check_arr_len(pubkey, 33)));
        const ret_hu_conv = new NodeId(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Create a new NodeId from a slice of bytes
     */
    static constructor_from_slice(bytes) {
        const ret = bindings.NodeId_from_slice(bindings.encodeUint8Array(bytes));
        const ret_hu_conv = Result_NodeIdDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Get the public key slice from this NodeId
     */
    as_slice() {
        const ret = bindings.NodeId_as_slice(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Get the public key as an array from this NodeId
     */
    as_array() {
        const ret = bindings.NodeId_as_array(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Get the public key from this NodeId
     */
    as_pubkey() {
        const ret = bindings.NodeId_as_pubkey(this.ptr);
        const ret_hu_conv = Result_PublicKeySecp256k1ErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Get the string representation of a NodeId object
     */
    to_str() {
        const ret = bindings.NodeId_to_str(this.ptr);
        const ret_conv = bindings.decodeString(ret);
        return ret_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the NodeId.
     */
    hash() {
        const ret = bindings.NodeId_hash(this.ptr);
        return ret;
    }
    /**
     * Serialize the NodeId object into a byte array which can be read by NodeId_read
     */
    write() {
        const ret = bindings.NodeId_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a NodeId from a byte array, created by NodeId_write
     */
    static constructor_read(ser) {
        const ret = bindings.NodeId_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_NodeIdDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
