// lightningdevkit/structs/OnionMessage.mjs
import { Result_OnionMessageDecodeErrorZ } from '../structs/Result_OnionMessageDecodeErrorZ.mjs';
import { Packet } from '../structs/Packet.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An onion message to be sent to or received from a peer.
 */
export class OnionMessage extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.OnionMessage_free);
    }
    /**
     * Used in decrypting the onion packet's payload.
     */
    get_blinding_point() {
        const ret = bindings.OnionMessage_get_blinding_point(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Used in decrypting the onion packet's payload.
     */
    set_blinding_point(val) {
        bindings.OnionMessage_set_blinding_point(this.ptr, bindings.encodeUint8Array(bindings.check_arr_len(val, 33)));
    }
    /**
     * The full onion packet including hop data, pubkey, and hmac
     */
    get_onion_routing_packet() {
        const ret = bindings.OnionMessage_get_onion_routing_packet(this.ptr);
        const ret_hu_conv = new Packet(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * The full onion packet including hop data, pubkey, and hmac
     */
    set_onion_routing_packet(val) {
        bindings.OnionMessage_set_onion_routing_packet(this.ptr, CommonBase.get_ptr_of(val));
    }
    /**
     * Constructs a new OnionMessage given each field
     */
    static constructor_new(blinding_point_arg, onion_routing_packet_arg) {
        const ret = bindings.OnionMessage_new(bindings.encodeUint8Array(bindings.check_arr_len(blinding_point_arg, 33)), CommonBase.get_ptr_of(onion_routing_packet_arg));
        const ret_hu_conv = new OnionMessage(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.OnionMessage_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the OnionMessage
     */
    clone() {
        const ret = bindings.OnionMessage_clone(this.ptr);
        const ret_hu_conv = new OnionMessage(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the OnionMessage.
     */
    hash() {
        const ret = bindings.OnionMessage_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two OnionMessages contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.OnionMessage_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Read a OnionMessage from a byte array, created by OnionMessage_write
     */
    static constructor_read(ser) {
        const ret = bindings.OnionMessage_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_OnionMessageDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Serialize the OnionMessage object into a byte array which can be read by OnionMessage_read
     */
    write() {
        const ret = bindings.OnionMessage_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
}
