// lightningdevkit/structs/Option_UtxoLookupZ.mjs
import { UtxoLookup } from '../structs/UtxoLookup.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::lightning::routing::utxo::UtxoLookup or not
 */
export class Option_UtxoLookupZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_UtxoLookupZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_UtxoLookupZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_UtxoLookupZ_Some(ptr);
            case 1: return new Option_UtxoLookupZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_UtxoLookupZ containing a crate::lightning::routing::utxo::UtxoLookup
     */
    static constructor_some(o) {
        const ret = bindings.COption_UtxoLookupZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_UtxoLookupZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        CommonBase.add_ref_from(ret_hu_conv, o);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_UtxoLookupZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_UtxoLookupZ_none();
        const ret_hu_conv = Option_UtxoLookupZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
}
/** A Option_UtxoLookupZ of type Some */
export class Option_UtxoLookupZ_Some extends Option_UtxoLookupZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_UtxoLookupZ_Some_get_some(ptr);
        const ret_hu_conv = new UtxoLookup(null, some);
        CommonBase.add_ref_from(ret_hu_conv, this);
        this.some = ret_hu_conv;
    }
}
/** A Option_UtxoLookupZ of type None */
export class Option_UtxoLookupZ_None extends Option_UtxoLookupZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
