// lightningdevkit/structs/Option_BigEndianScalarZ.mjs
import { BigEndianScalar } from '../structs/BigEndianScalar.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::BigEndianScalar or not
 */
export class Option_BigEndianScalarZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_BigEndianScalarZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_BigEndianScalarZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_BigEndianScalarZ_Some(ptr);
            case 1: return new Option_BigEndianScalarZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_BigEndianScalarZ containing a crate::c_types::BigEndianScalar
     */
    static constructor_some(o) {
        const ret = bindings.COption_BigEndianScalarZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_BigEndianScalarZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_BigEndianScalarZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_BigEndianScalarZ_none();
        const ret_hu_conv = Option_BigEndianScalarZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_BigEndianScalarZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_BigEndianScalarZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_BigEndianScalarZ_clone(this.ptr);
        const ret_hu_conv = Option_BigEndianScalarZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_BigEndianScalarZ of type Some */
export class Option_BigEndianScalarZ_Some extends Option_BigEndianScalarZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_BigEndianScalarZ_Some_get_some(ptr);
        const some_conv = new BigEndianScalar(null, some);
        this.some = some_conv;
    }
}
/** A Option_BigEndianScalarZ of type None */
export class Option_BigEndianScalarZ_None extends Option_BigEndianScalarZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
