// lightningdevkit/structs/Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ.mjs
import { TwoTuple_ReleaseHeldHtlcResponseInstructionZ } from '../structs/TwoTuple_ReleaseHeldHtlcResponseInstructionZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::derived::C2Tuple_ReleaseHeldHtlcResponseInstructionZ or not
 */
export class Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_Some(ptr);
            case 1: return new Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ containing a crate::c_types::derived::C2Tuple_ReleaseHeldHtlcResponseInstructionZ
     */
    static constructor_some(o) {
        const ret = bindings.COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_none();
        const ret_hu_conv = Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_clone(this.ptr);
        const ret_hu_conv = Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ of type Some */
export class Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_Some extends Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_Some_get_some(ptr);
        const some_hu_conv = new TwoTuple_ReleaseHeldHtlcResponseInstructionZ(null, some);
        CommonBase.add_ref_from(some_hu_conv, this);
        this.some = some_hu_conv;
    }
}
/** A Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ of type None */
export class Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ_None extends Option_C2Tuple_ReleaseHeldHtlcResponseInstructionZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
