// lightningdevkit/structs/Option_OnionMessageContentsZ.mjs
import { OnionMessageContents } from '../structs/OnionMessageContents.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::lightning::onion_message::packet::OnionMessageContents or not
 */
export class Option_OnionMessageContentsZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_OnionMessageContentsZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_OnionMessageContentsZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_OnionMessageContentsZ_Some(ptr);
            case 1: return new Option_OnionMessageContentsZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_OnionMessageContentsZ containing a crate::lightning::onion_message::packet::OnionMessageContents
     */
    static constructor_some(o) {
        const ret = bindings.COption_OnionMessageContentsZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_OnionMessageContentsZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        CommonBase.add_ref_from(ret_hu_conv, o);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_OnionMessageContentsZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_OnionMessageContentsZ_none();
        const ret_hu_conv = Option_OnionMessageContentsZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_OnionMessageContentsZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_OnionMessageContentsZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_OnionMessageContentsZ_clone(this.ptr);
        const ret_hu_conv = Option_OnionMessageContentsZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_OnionMessageContentsZ of type Some */
export class Option_OnionMessageContentsZ_Some extends Option_OnionMessageContentsZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_OnionMessageContentsZ_Some_get_some(ptr);
        const ret_hu_conv = new OnionMessageContents(null, some);
        CommonBase.add_ref_from(ret_hu_conv, this);
        this.some = ret_hu_conv;
    }
}
/** A Option_OnionMessageContentsZ of type None */
export class Option_OnionMessageContentsZ_None extends Option_OnionMessageContentsZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
