// lightningdevkit/structs/Option_C2Tuple_OffersMessageResponseInstructionZZ.mjs
import { TwoTuple_OffersMessageResponseInstructionZ } from '../structs/TwoTuple_OffersMessageResponseInstructionZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::derived::C2Tuple_OffersMessageResponseInstructionZ or not
 */
export class Option_C2Tuple_OffersMessageResponseInstructionZZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_C2Tuple_OffersMessageResponseInstructionZZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_C2Tuple_OffersMessageResponseInstructionZZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_C2Tuple_OffersMessageResponseInstructionZZ_Some(ptr);
            case 1: return new Option_C2Tuple_OffersMessageResponseInstructionZZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_C2Tuple_OffersMessageResponseInstructionZZ containing a crate::c_types::derived::C2Tuple_OffersMessageResponseInstructionZ
     */
    static constructor_some(o) {
        const ret = bindings.COption_C2Tuple_OffersMessageResponseInstructionZZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_C2Tuple_OffersMessageResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_C2Tuple_OffersMessageResponseInstructionZZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_C2Tuple_OffersMessageResponseInstructionZZ_none();
        const ret_hu_conv = Option_C2Tuple_OffersMessageResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_C2Tuple_OffersMessageResponseInstructionZZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_C2Tuple_OffersMessageResponseInstructionZZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_C2Tuple_OffersMessageResponseInstructionZZ_clone(this.ptr);
        const ret_hu_conv = Option_C2Tuple_OffersMessageResponseInstructionZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_C2Tuple_OffersMessageResponseInstructionZZ of type Some */
export class Option_C2Tuple_OffersMessageResponseInstructionZZ_Some extends Option_C2Tuple_OffersMessageResponseInstructionZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_C2Tuple_OffersMessageResponseInstructionZZ_Some_get_some(ptr);
        const some_hu_conv = new TwoTuple_OffersMessageResponseInstructionZ(null, some);
        CommonBase.add_ref_from(some_hu_conv, this);
        this.some = some_hu_conv;
    }
}
/** A Option_C2Tuple_OffersMessageResponseInstructionZZ of type None */
export class Option_C2Tuple_OffersMessageResponseInstructionZZ_None extends Option_C2Tuple_OffersMessageResponseInstructionZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
