// lightningdevkit/structs/Option_U128Z.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::U128 or not
 */
export class Option_U128Z extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_U128Z_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_U128Z_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_U128Z_Some(ptr);
            case 1: return new Option_U128Z_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_U128Z containing a crate::c_types::U128
     */
    static constructor_some(o) {
        const ret = bindings.COption_U128Z_some(bindings.encodeUint128(o));
        const ret_hu_conv = Option_U128Z.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_U128Z containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_U128Z_none();
        const ret_hu_conv = Option_U128Z.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_U128Z_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_U128Z which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_U128Z_clone(this.ptr);
        const ret_hu_conv = Option_U128Z.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_U128Z of type Some */
export class Option_U128Z_Some extends Option_U128Z {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_U128Z_Some_get_some(ptr);
        const some_conv = bindings.decodeUint128(some);
        this.some = some_conv;
    }
}
/** A Option_U128Z of type None */
export class Option_U128Z_None extends Option_U128Z {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
