// lightningdevkit/structs/Option_AmountZ.mjs
import { Amount } from '../structs/Amount.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::lightning::offers::offer::Amount or not
 */
export class Option_AmountZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_AmountZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_AmountZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_AmountZ_Some(ptr);
            case 1: return new Option_AmountZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_AmountZ containing a crate::lightning::offers::offer::Amount
     */
    static constructor_some(o) {
        const ret = bindings.COption_AmountZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_AmountZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_AmountZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_AmountZ_none();
        const ret_hu_conv = Option_AmountZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_AmountZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_AmountZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_AmountZ_clone(this.ptr);
        const ret_hu_conv = Option_AmountZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_AmountZ of type Some */
export class Option_AmountZ_Some extends Option_AmountZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_AmountZ_Some_get_some(ptr);
        const some_hu_conv = Amount.constr_from_ptr(some);
        CommonBase.add_ref_from(some_hu_conv, this);
        this.some = some_hu_conv;
    }
}
/** A Option_AmountZ of type None */
export class Option_AmountZ_None extends Option_AmountZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
