// lightningdevkit/structs/Option_CVec_u8ZZ.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::derived::CVec_u8Z or not
 */
export class Option_CVec_u8ZZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_CVec_u8ZZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_CVec_u8ZZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_CVec_u8ZZ_Some(ptr);
            case 1: return new Option_CVec_u8ZZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_CVec_u8ZZ containing a crate::c_types::derived::CVec_u8Z
     */
    static constructor_some(o) {
        const ret = bindings.COption_CVec_u8ZZ_some(bindings.encodeUint8Array(o));
        const ret_hu_conv = Option_CVec_u8ZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_CVec_u8ZZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_CVec_u8ZZ_none();
        const ret_hu_conv = Option_CVec_u8ZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_CVec_u8ZZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_CVec_u8ZZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_CVec_u8ZZ_clone(this.ptr);
        const ret_hu_conv = Option_CVec_u8ZZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_CVec_u8ZZ of type Some */
export class Option_CVec_u8ZZ_Some extends Option_CVec_u8ZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_CVec_u8ZZ_Some_get_some(ptr);
        const some_conv = bindings.decodeUint8Array(some);
        this.some = some_conv;
    }
}
/** A Option_CVec_u8ZZ of type None */
export class Option_CVec_u8ZZ_None extends Option_CVec_u8ZZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
