// lightningdevkit/structs/Option_ECDSASignatureZ.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::c_types::ECDSASignature or not
 */
export class Option_ECDSASignatureZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_ECDSASignatureZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_ECDSASignatureZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_ECDSASignatureZ_Some(ptr);
            case 1: return new Option_ECDSASignatureZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_ECDSASignatureZ containing a crate::c_types::ECDSASignature
     */
    static constructor_some(o) {
        const ret = bindings.COption_ECDSASignatureZ_some(bindings.encodeUint8Array(bindings.check_arr_len(o, 64)));
        const ret_hu_conv = Option_ECDSASignatureZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_ECDSASignatureZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_ECDSASignatureZ_none();
        const ret_hu_conv = Option_ECDSASignatureZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_ECDSASignatureZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_ECDSASignatureZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_ECDSASignatureZ_clone(this.ptr);
        const ret_hu_conv = Option_ECDSASignatureZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_ECDSASignatureZ of type Some */
export class Option_ECDSASignatureZ_Some extends Option_ECDSASignatureZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_ECDSASignatureZ_Some_get_some(ptr);
        const some_conv = bindings.decodeUint8Array(some);
        this.some = some_conv;
    }
}
/** A Option_ECDSASignatureZ of type None */
export class Option_ECDSASignatureZ_None extends Option_ECDSASignatureZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
