// lightningdevkit/structs/Option_MessageContextZ.mjs
import { MessageContext } from '../structs/MessageContext.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::lightning::blinded_path::message::MessageContext or not
 */
export class Option_MessageContextZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_MessageContextZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_MessageContextZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_MessageContextZ_Some(ptr);
            case 1: return new Option_MessageContextZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_MessageContextZ containing a crate::lightning::blinded_path::message::MessageContext
     */
    static constructor_some(o) {
        const ret = bindings.COption_MessageContextZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_MessageContextZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_MessageContextZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_MessageContextZ_none();
        const ret_hu_conv = Option_MessageContextZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_MessageContextZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_MessageContextZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_MessageContextZ_clone(this.ptr);
        const ret_hu_conv = Option_MessageContextZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_MessageContextZ of type Some */
export class Option_MessageContextZ_Some extends Option_MessageContextZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_MessageContextZ_Some_get_some(ptr);
        const some_hu_conv = MessageContext.constr_from_ptr(some);
        CommonBase.add_ref_from(some_hu_conv, this);
        this.some = some_hu_conv;
    }
}
/** A Option_MessageContextZ of type None */
export class Option_MessageContextZ_None extends Option_MessageContextZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
