// lightningdevkit/structs/Option_QuantityZ.mjs
import { Quantity } from '../structs/Quantity.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An enum which can either contain a crate::lightning::offers::offer::Quantity or not
 */
export class Option_QuantityZ extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.COption_QuantityZ_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKCOption_QuantityZ_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new Option_QuantityZ_Some(ptr);
            case 1: return new Option_QuantityZ_None(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    /**
     * Constructs a new COption_QuantityZ containing a crate::lightning::offers::offer::Quantity
     */
    static constructor_some(o) {
        const ret = bindings.COption_QuantityZ_some(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Option_QuantityZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Constructs a new COption_QuantityZ containing nothing
     */
    static constructor_none() {
        const ret = bindings.COption_QuantityZ_none();
        const ret_hu_conv = Option_QuantityZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.COption_QuantityZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new COption_QuantityZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.COption_QuantityZ_clone(this.ptr);
        const ret_hu_conv = Option_QuantityZ.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
/** A Option_QuantityZ of type Some */
export class Option_QuantityZ_Some extends Option_QuantityZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const some = bindings.LDKCOption_QuantityZ_Some_get_some(ptr);
        const some_hu_conv = Quantity.constr_from_ptr(some);
        CommonBase.add_ref_from(some_hu_conv, this);
        this.some = some_hu_conv;
    }
}
/** A Option_QuantityZ of type None */
export class Option_QuantityZ_None extends Option_QuantityZ {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
    }
}
