// lightningdevkit/structs/Result_BlindedFailureDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_BlindedFailureDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_BlindedFailureDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_BlindedFailureDecodeErrorZ_is_ok(ptr)) {
            return new Result_BlindedFailureDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_BlindedFailureDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_BlindedFailureDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_BlindedFailureDecodeErrorZ_ok(o);
        const ret_hu_conv = Result_BlindedFailureDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_BlindedFailureDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_BlindedFailureDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_BlindedFailureDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_BlindedFailureDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_BlindedFailureDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_BlindedFailureDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_BlindedFailureDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_BlindedFailureDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_BlindedFailureDecodeErrorZ_OK extends Result_BlindedFailureDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        this.res = bindings.CResult_BlindedFailureDecodeErrorZ_get_ok(ptr);
    }
}
export class Result_BlindedFailureDecodeErrorZ_Err extends Result_BlindedFailureDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_BlindedFailureDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
