// lightningdevkit/structs/Result_PendingHTLCRoutingDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { PendingHTLCRouting } from '../structs/PendingHTLCRouting.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_PendingHTLCRoutingDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_PendingHTLCRoutingDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_PendingHTLCRoutingDecodeErrorZ_is_ok(ptr)) {
            return new Result_PendingHTLCRoutingDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_PendingHTLCRoutingDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_PendingHTLCRoutingDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_PendingHTLCRoutingDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_PendingHTLCRoutingDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_PendingHTLCRoutingDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_PendingHTLCRoutingDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_PendingHTLCRoutingDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_PendingHTLCRoutingDecodeErrorZ_OK extends Result_PendingHTLCRoutingDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_get_ok(ptr);
        const res_hu_conv = PendingHTLCRouting.constr_from_ptr(res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_PendingHTLCRoutingDecodeErrorZ_Err extends Result_PendingHTLCRoutingDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_PendingHTLCRoutingDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
