// lightningdevkit/structs/Result_SchnorrSignatureNoneZ.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_SchnorrSignatureNoneZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_SchnorrSignatureNoneZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_SchnorrSignatureNoneZ_is_ok(ptr)) {
            return new Result_SchnorrSignatureNoneZ_OK(null, ptr);
        }
        else {
            return new Result_SchnorrSignatureNoneZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_SchnorrSignatureNoneZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_SchnorrSignatureNoneZ_ok(bindings.encodeUint8Array(bindings.check_arr_len(o, 64)));
        const ret_hu_conv = Result_SchnorrSignatureNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_SchnorrSignatureNoneZ in the error state.
     */
    static constructor_err() {
        const ret = bindings.CResult_SchnorrSignatureNoneZ_err();
        const ret_hu_conv = Result_SchnorrSignatureNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_SchnorrSignatureNoneZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_SchnorrSignatureNoneZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_SchnorrSignatureNoneZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_SchnorrSignatureNoneZ_clone(this.ptr);
        const ret_hu_conv = Result_SchnorrSignatureNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_SchnorrSignatureNoneZ_OK extends Result_SchnorrSignatureNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_SchnorrSignatureNoneZ_get_ok(ptr);
        const res_conv = bindings.decodeUint8Array(res);
        this.res = res_conv;
    }
}
export class Result_SchnorrSignatureNoneZ_Err extends Result_SchnorrSignatureNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
