// lightningdevkit/structs/Result_NoneBolt11SemanticErrorZ.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_NoneBolt11SemanticErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_NoneBolt11SemanticErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_NoneBolt11SemanticErrorZ_is_ok(ptr)) {
            return new Result_NoneBolt11SemanticErrorZ_OK(null, ptr);
        }
        else {
            return new Result_NoneBolt11SemanticErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_NoneBolt11SemanticErrorZ in the success state.
     */
    static constructor_ok() {
        const ret = bindings.CResult_NoneBolt11SemanticErrorZ_ok();
        const ret_hu_conv = Result_NoneBolt11SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_NoneBolt11SemanticErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_NoneBolt11SemanticErrorZ_err(e);
        const ret_hu_conv = Result_NoneBolt11SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_NoneBolt11SemanticErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_NoneBolt11SemanticErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_NoneBolt11SemanticErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_NoneBolt11SemanticErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_NoneBolt11SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_NoneBolt11SemanticErrorZ_OK extends Result_NoneBolt11SemanticErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
export class Result_NoneBolt11SemanticErrorZ_Err extends Result_NoneBolt11SemanticErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        this.err = bindings.CResult_NoneBolt11SemanticErrorZ_get_err(ptr);
    }
}
