// lightningdevkit/structs/Result_TxAckRbfDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { TxAckRbf } from '../structs/TxAckRbf.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_TxAckRbfDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_TxAckRbfDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_TxAckRbfDecodeErrorZ_is_ok(ptr)) {
            return new Result_TxAckRbfDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_TxAckRbfDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_TxAckRbfDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_TxAckRbfDecodeErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_TxAckRbfDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_TxAckRbfDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_TxAckRbfDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_TxAckRbfDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_TxAckRbfDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_TxAckRbfDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_TxAckRbfDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_TxAckRbfDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_TxAckRbfDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_TxAckRbfDecodeErrorZ_OK extends Result_TxAckRbfDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_TxAckRbfDecodeErrorZ_get_ok(ptr);
        const res_hu_conv = new TxAckRbf(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_TxAckRbfDecodeErrorZ_Err extends Result_TxAckRbfDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_TxAckRbfDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
