// lightningdevkit/structs/Result_SpliceAckDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { SpliceAck } from '../structs/SpliceAck.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_SpliceAckDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_SpliceAckDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_SpliceAckDecodeErrorZ_is_ok(ptr)) {
            return new Result_SpliceAckDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_SpliceAckDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_SpliceAckDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_SpliceAckDecodeErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_SpliceAckDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_SpliceAckDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_SpliceAckDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_SpliceAckDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_SpliceAckDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_SpliceAckDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_SpliceAckDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_SpliceAckDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_SpliceAckDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_SpliceAckDecodeErrorZ_OK extends Result_SpliceAckDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_SpliceAckDecodeErrorZ_get_ok(ptr);
        const res_hu_conv = new SpliceAck(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_SpliceAckDecodeErrorZ_Err extends Result_SpliceAckDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_SpliceAckDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
