// lightningdevkit/structs/Result_OffersMessageDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { OffersMessage } from '../structs/OffersMessage.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_OffersMessageDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_OffersMessageDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_OffersMessageDecodeErrorZ_is_ok(ptr)) {
            return new Result_OffersMessageDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_OffersMessageDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_OffersMessageDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_OffersMessageDecodeErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_OffersMessageDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_OffersMessageDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_OffersMessageDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_OffersMessageDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_OffersMessageDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_OffersMessageDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_OffersMessageDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_OffersMessageDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_OffersMessageDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_OffersMessageDecodeErrorZ_OK extends Result_OffersMessageDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_OffersMessageDecodeErrorZ_get_ok(ptr);
        const res_hu_conv = OffersMessage.constr_from_ptr(res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_OffersMessageDecodeErrorZ_Err extends Result_OffersMessageDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_OffersMessageDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
