// lightningdevkit/structs/Result_TransactionU16LenLimitedNoneZ.mjs
import { TransactionU16LenLimited } from '../structs/TransactionU16LenLimited.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_TransactionU16LenLimitedNoneZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_TransactionU16LenLimitedNoneZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_TransactionU16LenLimitedNoneZ_is_ok(ptr)) {
            return new Result_TransactionU16LenLimitedNoneZ_OK(null, ptr);
        }
        else {
            return new Result_TransactionU16LenLimitedNoneZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedNoneZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_TransactionU16LenLimitedNoneZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_TransactionU16LenLimitedNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedNoneZ in the error state.
     */
    static constructor_err() {
        const ret = bindings.CResult_TransactionU16LenLimitedNoneZ_err();
        const ret_hu_conv = Result_TransactionU16LenLimitedNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_TransactionU16LenLimitedNoneZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_TransactionU16LenLimitedNoneZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedNoneZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_TransactionU16LenLimitedNoneZ_clone(this.ptr);
        const ret_hu_conv = Result_TransactionU16LenLimitedNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_TransactionU16LenLimitedNoneZ_OK extends Result_TransactionU16LenLimitedNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_TransactionU16LenLimitedNoneZ_get_ok(ptr);
        const res_hu_conv = new TransactionU16LenLimited(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_TransactionU16LenLimitedNoneZ_Err extends Result_TransactionU16LenLimitedNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
