// lightningdevkit/structs/Result_RecipientOnionFieldsNoneZ.mjs
import { RecipientOnionFields } from '../structs/RecipientOnionFields.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_RecipientOnionFieldsNoneZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_RecipientOnionFieldsNoneZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_RecipientOnionFieldsNoneZ_is_ok(ptr)) {
            return new Result_RecipientOnionFieldsNoneZ_OK(null, ptr);
        }
        else {
            return new Result_RecipientOnionFieldsNoneZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_RecipientOnionFieldsNoneZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_RecipientOnionFieldsNoneZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_RecipientOnionFieldsNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_RecipientOnionFieldsNoneZ in the error state.
     */
    static constructor_err() {
        const ret = bindings.CResult_RecipientOnionFieldsNoneZ_err();
        const ret_hu_conv = Result_RecipientOnionFieldsNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_RecipientOnionFieldsNoneZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_RecipientOnionFieldsNoneZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_RecipientOnionFieldsNoneZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_RecipientOnionFieldsNoneZ_clone(this.ptr);
        const ret_hu_conv = Result_RecipientOnionFieldsNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_RecipientOnionFieldsNoneZ_OK extends Result_RecipientOnionFieldsNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_RecipientOnionFieldsNoneZ_get_ok(ptr);
        const res_hu_conv = new RecipientOnionFields(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_RecipientOnionFieldsNoneZ_Err extends Result_RecipientOnionFieldsNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
