// lightningdevkit/structs/Result_ChannelMonitorUpdateStatusNoneZ.mjs
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_ChannelMonitorUpdateStatusNoneZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_ChannelMonitorUpdateStatusNoneZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_ChannelMonitorUpdateStatusNoneZ_is_ok(ptr)) {
            return new Result_ChannelMonitorUpdateStatusNoneZ_OK(null, ptr);
        }
        else {
            return new Result_ChannelMonitorUpdateStatusNoneZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_ChannelMonitorUpdateStatusNoneZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_ok(o);
        const ret_hu_conv = Result_ChannelMonitorUpdateStatusNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_ChannelMonitorUpdateStatusNoneZ in the error state.
     */
    static constructor_err() {
        const ret = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_err();
        const ret_hu_conv = Result_ChannelMonitorUpdateStatusNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_ChannelMonitorUpdateStatusNoneZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_clone(this.ptr);
        const ret_hu_conv = Result_ChannelMonitorUpdateStatusNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_ChannelMonitorUpdateStatusNoneZ_OK extends Result_ChannelMonitorUpdateStatusNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        this.res = bindings.CResult_ChannelMonitorUpdateStatusNoneZ_get_ok(ptr);
    }
}
export class Result_ChannelMonitorUpdateStatusNoneZ_Err extends Result_ChannelMonitorUpdateStatusNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
