// lightningdevkit/structs/Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ.mjs
import { UnsignedBolt12Invoice } from '../structs/UnsignedBolt12Invoice.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_is_ok(ptr)) {
            return new Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ_OK(null, ptr);
        }
        else {
            return new Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_err(e);
        const ret_hu_conv = Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ_OK extends Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_get_ok(ptr);
        const res_hu_conv = new UnsignedBolt12Invoice(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ_Err extends Result_UnsignedBolt12InvoiceBolt12SemanticErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        this.err = bindings.CResult_UnsignedBolt12InvoiceBolt12SemanticErrorZ_get_err(ptr);
    }
}
