// lightningdevkit/structs/Result_SignedRawBolt11InvoiceBolt11ParseErrorZ.mjs
import { Bolt11ParseError } from '../structs/Bolt11ParseError.mjs';
import { SignedRawBolt11Invoice } from '../structs/SignedRawBolt11Invoice.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_SignedRawBolt11InvoiceBolt11ParseErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_is_ok(ptr)) {
            return new Result_SignedRawBolt11InvoiceBolt11ParseErrorZ_OK(null, ptr);
        }
        else {
            return new Result_SignedRawBolt11InvoiceBolt11ParseErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_SignedRawBolt11InvoiceBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_SignedRawBolt11InvoiceBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_SignedRawBolt11InvoiceBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_SignedRawBolt11InvoiceBolt11ParseErrorZ_OK extends Result_SignedRawBolt11InvoiceBolt11ParseErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_get_ok(ptr);
        const res_hu_conv = new SignedRawBolt11Invoice(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_SignedRawBolt11InvoiceBolt11ParseErrorZ_Err extends Result_SignedRawBolt11InvoiceBolt11ParseErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_SignedRawBolt11InvoiceBolt11ParseErrorZ_get_err(ptr);
        const err_hu_conv = Bolt11ParseError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
