// lightningdevkit/structs/Result_SiPrefixBolt11ParseErrorZ.mjs
import { Bolt11ParseError } from '../structs/Bolt11ParseError.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_SiPrefixBolt11ParseErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_SiPrefixBolt11ParseErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_SiPrefixBolt11ParseErrorZ_is_ok(ptr)) {
            return new Result_SiPrefixBolt11ParseErrorZ_OK(null, ptr);
        }
        else {
            return new Result_SiPrefixBolt11ParseErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_SiPrefixBolt11ParseErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_SiPrefixBolt11ParseErrorZ_ok(o);
        const ret_hu_conv = Result_SiPrefixBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_SiPrefixBolt11ParseErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_SiPrefixBolt11ParseErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_SiPrefixBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_SiPrefixBolt11ParseErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_SiPrefixBolt11ParseErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_SiPrefixBolt11ParseErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_SiPrefixBolt11ParseErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_SiPrefixBolt11ParseErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_SiPrefixBolt11ParseErrorZ_OK extends Result_SiPrefixBolt11ParseErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        this.res = bindings.CResult_SiPrefixBolt11ParseErrorZ_get_ok(ptr);
    }
}
export class Result_SiPrefixBolt11ParseErrorZ_Err extends Result_SiPrefixBolt11ParseErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_SiPrefixBolt11ParseErrorZ_get_err(ptr);
        const err_hu_conv = Bolt11ParseError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
