// lightningdevkit/structs/Result_TransactionU16LenLimitedDecodeErrorZ.mjs
import { DecodeError } from '../structs/DecodeError.mjs';
import { TransactionU16LenLimited } from '../structs/TransactionU16LenLimited.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_TransactionU16LenLimitedDecodeErrorZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_is_ok(ptr)) {
            return new Result_TransactionU16LenLimitedDecodeErrorZ_OK(null, ptr);
        }
        else {
            return new Result_TransactionU16LenLimitedDecodeErrorZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedDecodeErrorZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_TransactionU16LenLimitedDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedDecodeErrorZ in the error state.
     */
    static constructor_err(e) {
        const ret = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_err(CommonBase.get_ptr_of(e));
        const ret_hu_conv = Result_TransactionU16LenLimitedDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_TransactionU16LenLimitedDecodeErrorZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_clone(this.ptr);
        const ret_hu_conv = Result_TransactionU16LenLimitedDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_TransactionU16LenLimitedDecodeErrorZ_OK extends Result_TransactionU16LenLimitedDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_get_ok(ptr);
        const res_hu_conv = new TransactionU16LenLimited(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_TransactionU16LenLimitedDecodeErrorZ_Err extends Result_TransactionU16LenLimitedDecodeErrorZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const err = bindings.CResult_TransactionU16LenLimitedDecodeErrorZ_get_err(ptr);
        const err_hu_conv = DecodeError.constr_from_ptr(err);
        CommonBase.add_ref_from(err_hu_conv, this);
        this.err = err_hu_conv;
    }
}
