// lightningdevkit/structs/Result_OnionMessagePathNoneZ.mjs
import { OnionMessagePath } from '../structs/OnionMessagePath.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
export class Result_OnionMessagePathNoneZ extends CommonBase {
    constructor(_dummy, ptr) {
        super(ptr, bindings.CResult_OnionMessagePathNoneZ_free);
    }
    /* @internal */
    static constr_from_ptr(ptr) {
        if (bindings.CResult_OnionMessagePathNoneZ_is_ok(ptr)) {
            return new Result_OnionMessagePathNoneZ_OK(null, ptr);
        }
        else {
            return new Result_OnionMessagePathNoneZ_Err(null, ptr);
        }
    }
    /**
     * Creates a new CResult_OnionMessagePathNoneZ in the success state.
     */
    static constructor_ok(o) {
        const ret = bindings.CResult_OnionMessagePathNoneZ_ok(CommonBase.get_ptr_of(o));
        const ret_hu_conv = Result_OnionMessagePathNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Creates a new CResult_OnionMessagePathNoneZ in the error state.
     */
    static constructor_err() {
        const ret = bindings.CResult_OnionMessagePathNoneZ_err();
        const ret_hu_conv = Result_OnionMessagePathNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Checks if the given object is currently in the success state
     */
    is_ok() {
        const ret = bindings.CResult_OnionMessagePathNoneZ_is_ok(this.ptr);
        return ret;
    }
    clone_ptr() {
        const ret = bindings.CResult_OnionMessagePathNoneZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new CResult_OnionMessagePathNoneZ which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.CResult_OnionMessagePathNoneZ_clone(this.ptr);
        const ret_hu_conv = Result_OnionMessagePathNoneZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
export class Result_OnionMessagePathNoneZ_OK extends Result_OnionMessagePathNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
        const res = bindings.CResult_OnionMessagePathNoneZ_get_ok(ptr);
        const res_hu_conv = new OnionMessagePath(null, res);
        CommonBase.add_ref_from(res_hu_conv, this);
        this.res = res_hu_conv;
    }
}
export class Result_OnionMessagePathNoneZ_Err extends Result_OnionMessagePathNoneZ {
    /* @internal */
    constructor(_dummy, ptr) {
        super(_dummy, ptr);
    }
}
