// lightningdevkit/structs/SocketAddress.mjs
import { Hostname } from '../structs/Hostname.mjs';
import { Result_SocketAddressDecodeErrorZ } from '../structs/Result_SocketAddressDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An address which can be used to connect to a remote peer.
 */
export class SocketAddress extends CommonBase {
    constructor(_dummy, ptr) { super(ptr, bindings.SocketAddress_free); }
    /* @internal */
    static constr_from_ptr(ptr) {
        const raw_ty = bindings.LDKSocketAddress_ty_from_ptr(ptr);
        switch (raw_ty) {
            case 0: return new SocketAddress_TcpIpV4(ptr);
            case 1: return new SocketAddress_TcpIpV6(ptr);
            case 2: return new SocketAddress_OnionV2(ptr);
            case 3: return new SocketAddress_OnionV3(ptr);
            case 4: return new SocketAddress_Hostname(ptr);
            default:
                throw new Error('oops, this should be unreachable'); // Unreachable without extending the (internal) bindings interface
        }
    }
    clone_ptr() {
        const ret = bindings.SocketAddress_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the SocketAddress
     */
    clone() {
        const ret = bindings.SocketAddress_clone(this.ptr);
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new TcpIpV4-variant SocketAddress
     */
    static constructor_tcp_ip_v4(addr, port) {
        const ret = bindings.SocketAddress_tcp_ip_v4(bindings.encodeUint8Array(bindings.check_arr_len(addr, 4)), port);
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new TcpIpV6-variant SocketAddress
     */
    static constructor_tcp_ip_v6(addr, port) {
        const ret = bindings.SocketAddress_tcp_ip_v6(bindings.encodeUint8Array(bindings.check_arr_len(addr, 16)), port);
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new OnionV2-variant SocketAddress
     */
    static constructor_onion_v2(a) {
        const ret = bindings.SocketAddress_onion_v2(bindings.encodeUint8Array(bindings.check_arr_len(a, 12)));
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new OnionV3-variant SocketAddress
     */
    static constructor_onion_v3(ed25519_pubkey, checksum, version, port) {
        const ret = bindings.SocketAddress_onion_v3(bindings.encodeUint8Array(bindings.check_arr_len(ed25519_pubkey, 32)), checksum, version, port);
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Utility method to constructs a new Hostname-variant SocketAddress
     */
    static constructor_hostname(hostname, port) {
        const ret = bindings.SocketAddress_hostname(CommonBase.get_ptr_of(hostname), port);
        const ret_hu_conv = SocketAddress.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the SocketAddress.
     */
    hash() {
        const ret = bindings.SocketAddress_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two SocketAddresss contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     */
    eq(b) {
        const ret = bindings.SocketAddress_eq(this.ptr, CommonBase.get_ptr_of(b));
        return ret;
    }
    /**
     * Serialize the SocketAddress object into a byte array which can be read by SocketAddress_read
     */
    write() {
        const ret = bindings.SocketAddress_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a SocketAddress from a byte array, created by SocketAddress_write
     */
    static constructor_read(ser) {
        const ret = bindings.SocketAddress_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_SocketAddressDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
    /**
     * Get the string representation of a SocketAddress object
     */
    to_str() {
        const ret = bindings.SocketAddress_to_str(this.ptr);
        const ret_conv = bindings.decodeString(ret);
        return ret_conv;
    }
}
/** A SocketAddress of type TcpIpV4 */
export class SocketAddress_TcpIpV4 extends SocketAddress {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const addr = bindings.LDKSocketAddress_TcpIpV4_get_addr(ptr);
        const addr_conv = bindings.decodeUint8Array(addr);
        this.addr = addr_conv;
        this.port = bindings.LDKSocketAddress_TcpIpV4_get_port(ptr);
    }
}
/** A SocketAddress of type TcpIpV6 */
export class SocketAddress_TcpIpV6 extends SocketAddress {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const addr = bindings.LDKSocketAddress_TcpIpV6_get_addr(ptr);
        const addr_conv = bindings.decodeUint8Array(addr);
        this.addr = addr_conv;
        this.port = bindings.LDKSocketAddress_TcpIpV6_get_port(ptr);
    }
}
/** A SocketAddress of type OnionV2 */
export class SocketAddress_OnionV2 extends SocketAddress {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const onion_v2 = bindings.LDKSocketAddress_OnionV2_get_onion_v2(ptr);
        const onion_v2_conv = bindings.decodeUint8Array(onion_v2);
        this.onion_v2 = onion_v2_conv;
    }
}
/** A SocketAddress of type OnionV3 */
export class SocketAddress_OnionV3 extends SocketAddress {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const ed25519_pubkey = bindings.LDKSocketAddress_OnionV3_get_ed25519_pubkey(ptr);
        const ed25519_pubkey_conv = bindings.decodeUint8Array(ed25519_pubkey);
        this.ed25519_pubkey = ed25519_pubkey_conv;
        this.checksum = bindings.LDKSocketAddress_OnionV3_get_checksum(ptr);
        this.version = bindings.LDKSocketAddress_OnionV3_get_version(ptr);
        this.port = bindings.LDKSocketAddress_OnionV3_get_port(ptr);
    }
}
/** A SocketAddress of type Hostname */
export class SocketAddress_Hostname extends SocketAddress {
    /* @internal */
    constructor(ptr) {
        super(null, ptr);
        const hostname = bindings.LDKSocketAddress_Hostname_get_hostname(ptr);
        const hostname_hu_conv = new Hostname(null, hostname);
        CommonBase.add_ref_from(hostname_hu_conv, this);
        this.hostname = hostname_hu_conv;
        this.port = bindings.LDKSocketAddress_Hostname_get_port(ptr);
    }
}
