// lightningdevkit/structs/TwoTuple_ReleaseHeldHtlcResponseInstructionZ.mjs
import { ResponseInstruction } from '../structs/ResponseInstruction.mjs';
import { ReleaseHeldHtlc } from '../structs/ReleaseHeldHtlc.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * A Tuple
 */
export class TwoTuple_ReleaseHeldHtlcResponseInstructionZ extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_free);
    }
    /**
     *
     */
    get_a() {
        const ret = bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_get_a(this.ptr);
        const ret_hu_conv = new ReleaseHeldHtlc(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     *
     */
    get_b() {
        const ret = bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_get_b(this.ptr);
        const ret_hu_conv = new ResponseInstruction(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new tuple which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_clone(this.ptr);
        const ret_hu_conv = new TwoTuple_ReleaseHeldHtlcResponseInstructionZ(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Creates a new C2Tuple_ReleaseHeldHtlcResponseInstructionZ from the contained elements.
     */
    static constructor_new(a, b) {
        const ret = bindings.C2Tuple_ReleaseHeldHtlcResponseInstructionZ_new(CommonBase.get_ptr_of(a), CommonBase.get_ptr_of(b));
        const ret_hu_conv = new TwoTuple_ReleaseHeldHtlcResponseInstructionZ(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
}
