// lightningdevkit/structs/TwoTuple_OffersMessageMessageSendInstructionsZ.mjs
import { OffersMessage } from '../structs/OffersMessage.mjs';
import { MessageSendInstructions } from '../structs/MessageSendInstructions.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * A Tuple
 */
export class TwoTuple_OffersMessageMessageSendInstructionsZ extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_free);
    }
    /**
     *
     */
    get_a() {
        const ret = bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_get_a(this.ptr);
        const ret_hu_conv = OffersMessage.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     *
     */
    get_b() {
        const ret = bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_get_b(this.ptr);
        const ret_hu_conv = MessageSendInstructions.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a new tuple which has the same data as `orig`
     * but with all dynamically-allocated buffers duplicated in new buffers.
     */
    clone() {
        const ret = bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_clone(this.ptr);
        const ret_hu_conv = new TwoTuple_OffersMessageMessageSendInstructionsZ(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Creates a new C2Tuple_OffersMessageMessageSendInstructionsZ from the contained elements.
     */
    static constructor_new(a, b) {
        const ret = bindings.C2Tuple_OffersMessageMessageSendInstructionsZ_new(CommonBase.get_ptr_of(a), CommonBase.get_ptr_of(b));
        const ret_hu_conv = new TwoTuple_OffersMessageMessageSendInstructionsZ(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
}
