// lightningdevkit/structs/TxRemoveOutput.mjs
import { ChannelId } from '../structs/ChannelId.mjs';
import { Result_TxRemoveOutputDecodeErrorZ } from '../structs/Result_TxRemoveOutputDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * A tx_remove_output message for removing an output during interactive transaction construction.
 */
export class TxRemoveOutput extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.TxRemoveOutput_free);
    }
    /**
     * The channel ID
     */
    get_channel_id() {
        const ret = bindings.TxRemoveOutput_get_channel_id(this.ptr);
        const ret_hu_conv = new ChannelId(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * The channel ID
     */
    set_channel_id(val) {
        bindings.TxRemoveOutput_set_channel_id(this.ptr, CommonBase.get_ptr_of(val));
    }
    /**
     * The serial ID of the output to be removed
     */
    get_serial_id() {
        const ret = bindings.TxRemoveOutput_get_serial_id(this.ptr);
        return ret;
    }
    /**
     * The serial ID of the output to be removed
     */
    set_serial_id(val) {
        bindings.TxRemoveOutput_set_serial_id(this.ptr, val);
    }
    /**
     * Constructs a new TxRemoveOutput given each field
     */
    static constructor_new(channel_id_arg, serial_id_arg) {
        const ret = bindings.TxRemoveOutput_new(CommonBase.get_ptr_of(channel_id_arg), serial_id_arg);
        const ret_hu_conv = new TxRemoveOutput(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, ret_hu_conv);
        return ret_hu_conv;
    }
    clone_ptr() {
        const ret = bindings.TxRemoveOutput_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the TxRemoveOutput
     */
    clone() {
        const ret = bindings.TxRemoveOutput_clone(this.ptr);
        const ret_hu_conv = new TxRemoveOutput(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Generates a non-cryptographic 64-bit hash of the TxRemoveOutput.
     */
    hash() {
        const ret = bindings.TxRemoveOutput_hash(this.ptr);
        return ret;
    }
    /**
     * Checks if two TxRemoveOutputs contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.TxRemoveOutput_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Serialize the TxRemoveOutput object into a byte array which can be read by TxRemoveOutput_read
     */
    write() {
        const ret = bindings.TxRemoveOutput_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a TxRemoveOutput from a byte array, created by TxRemoveOutput_write
     */
    static constructor_read(ser) {
        const ret = bindings.TxRemoveOutput_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_TxRemoveOutputDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
