// lightningdevkit/structs/UnknownPaymentContext.mjs
import { Result_UnknownPaymentContextDecodeErrorZ } from '../structs/Result_UnknownPaymentContextDecodeErrorZ.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
/**
 * An unknown payment context.
 */
export class UnknownPaymentContext extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.UnknownPaymentContext_free);
    }
    clone_ptr() {
        const ret = bindings.UnknownPaymentContext_clone_ptr(this.ptr);
        return ret;
    }
    /**
     * Creates a copy of the UnknownPaymentContext
     */
    clone() {
        const ret = bindings.UnknownPaymentContext_clone(this.ptr);
        const ret_hu_conv = new UnknownPaymentContext(null, ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
    /**
     * Checks if two UnknownPaymentContexts contain equal inner contents.
     * This ignores pointers and is_owned flags and looks at the values in fields.
     * Two objects with NULL inner values will be considered "equal" here.
     */
    eq(b) {
        const ret = bindings.UnknownPaymentContext_eq(this.ptr, CommonBase.get_ptr_of(b));
        CommonBase.add_ref_from(this, b);
        return ret;
    }
    /**
     * Serialize the UnknownPaymentContext object into a byte array which can be read by UnknownPaymentContext_read
     */
    write() {
        const ret = bindings.UnknownPaymentContext_write(this.ptr);
        const ret_conv = bindings.decodeUint8Array(ret);
        return ret_conv;
    }
    /**
     * Read a UnknownPaymentContext from a byte array, created by UnknownPaymentContext_write
     */
    static constructor_read(ser) {
        const ret = bindings.UnknownPaymentContext_read(bindings.encodeUint8Array(ser));
        const ret_hu_conv = Result_UnknownPaymentContextDecodeErrorZ.constr_from_ptr(ret);
        return ret_hu_conv;
    }
}
