// lightningdevkit/structs/UtxoLookup.mjs
import { UtxoResult } from '../structs/UtxoResult.mjs';
import { CommonBase } from './CommonBase.mjs';
import * as bindings from '../bindings.mjs';
class LDKUtxoLookupHolder {
    constructor() {
        this.held = null;
    }
}
/**
 * The `UtxoLookup` trait defines behavior for accessing on-chain UTXOs.
 */
export class UtxoLookup extends CommonBase {
    /* @internal */
    constructor(_dummy, ptr) {
        super(ptr, bindings.UtxoLookup_free);
        this.bindings_instance = null;
    }
    /** Creates a new instance of UtxoLookup from a given implementation */
    static new_impl(arg) {
        const impl_holder = new LDKUtxoLookupHolder();
        let structImplementation = {
            get_utxo(chain_hash, short_channel_id) {
                const chain_hash_conv = bindings.decodeUint8Array(chain_hash);
                const ret = arg.get_utxo(chain_hash_conv, short_channel_id);
                const result = ret.clone_ptr();
                return result;
            },
        };
        const ptr_idx = bindings.LDKUtxoLookup_new(structImplementation);
        impl_holder.held = new UtxoLookup(null, ptr_idx[0]);
        impl_holder.held.instance_idx = ptr_idx[1];
        impl_holder.held.bindings_instance = structImplementation;
        return impl_holder.held;
    }
    /**
     * Returns the transaction output of a funding transaction encoded by [`short_channel_id`].
     * Returns an error if `chain_hash` is for a different chain or if such a transaction output is
     * unknown.
     *
     * [`short_channel_id`]: https://github.com/lightning/bolts/blob/master/07-routing-gossip.md#definition-of-short_channel_id
     */
    get_utxo(chain_hash, short_channel_id) {
        const ret = bindings.UtxoLookup_get_utxo(this.ptr, bindings.encodeUint8Array(bindings.check_arr_len(chain_hash, 32)), short_channel_id);
        const ret_hu_conv = UtxoResult.constr_from_ptr(ret);
        CommonBase.add_ref_from(ret_hu_conv, this);
        return ret_hu_conv;
    }
}
