// @synonymdev/react-native-ldk/dist/utils/types.js
export var ENetworks;
(function (ENetworks) {
    ENetworks["signet"] = "signet";
    ENetworks["regtest"] = "regtest";
    ENetworks["testnet"] = "testnet";
    ENetworks["mainnet"] = "bitcoin";
})(ENetworks || (ENetworks = {}));
export var EEventTypes;
(function (EEventTypes) {
    EEventTypes["ldk_log"] = "ldk_log";
    EEventTypes["native_log"] = "native_log";
    EEventTypes["register_tx"] = "register_tx";
    EEventTypes["register_output"] = "register_output";
    EEventTypes["broadcast_transaction"] = "broadcast_transaction";
    EEventTypes["channel_manager_funding_generation_ready"] = "channel_manager_funding_generation_ready";
    EEventTypes["channel_manager_payment_claimable"] = "channel_manager_payment_claimable";
    EEventTypes["channel_manager_payment_sent"] = "channel_manager_payment_sent";
    EEventTypes["channel_manager_open_channel_request"] = "channel_manager_open_channel_request";
    EEventTypes["channel_manager_payment_path_successful"] = "channel_manager_payment_path_successful";
    EEventTypes["channel_manager_payment_path_failed"] = "channel_manager_payment_path_failed";
    EEventTypes["channel_manager_payment_failed"] = "channel_manager_payment_failed";
    EEventTypes["channel_manager_pending_htlcs_forwardable"] = "channel_manager_pending_htlcs_forwardable";
    EEventTypes["channel_manager_spendable_outputs"] = "channel_manager_spendable_outputs";
    EEventTypes["channel_manager_channel_closed"] = "channel_manager_channel_closed";
    EEventTypes["channel_manager_discard_funding"] = "channel_manager_discard_funding";
    EEventTypes["channel_manager_payment_claimed"] = "channel_manager_payment_claimed";
    EEventTypes["emergency_force_close_channel"] = "emergency_force_close_channel";
    EEventTypes["new_channel"] = "new_channel";
    EEventTypes["network_graph_updated"] = "network_graph_updated";
    EEventTypes["channel_manager_restarted"] = "channel_manager_restarted";
    EEventTypes["backup_state_update"] = "backup_state_update";
    EEventTypes["lsp_log"] = "lsp_log";
    EEventTypes["used_close_address"] = "used_close_address";
})(EEventTypes || (EEventTypes = {}));
export const defaultUserConfig = {
    channel_handshake_config: {
        announced_channel: false,
        minimum_depth: 1,
        max_htlc_value_in_flight_percent_of_channel: 100,
        max_inbound_htlc_value_in_flight_percent_of_channel: 100,
        negotiate_anchors_zero_fee_htlc_tx: true,
    },
    manually_accept_inbound_channels: true,
    accept_inbound_channels: true,
};
export var ELdkLogLevels;
(function (ELdkLogLevels) {
    ELdkLogLevels["gossip"] = "GOSSIP";
    ELdkLogLevels["trace"] = "TRACE";
    ELdkLogLevels["debug"] = "DEBUG";
    ELdkLogLevels["info"] = "INFO";
    ELdkLogLevels["warn"] = "WARN";
    ELdkLogLevels["error"] = "ERROR";
})(ELdkLogLevels || (ELdkLogLevels = {}));
export const DefaultTransactionDataShape = {
    header: '',
    height: 0,
    transaction: '',
    vout: [],
};
export var ELdkFiles;
(function (ELdkFiles) {
    ELdkFiles["seed"] = "seed";
    ELdkFiles["channel_manager"] = "channel_manager.bin";
    ELdkFiles["channels"] = "channels";
    ELdkFiles["peers"] = "peers.json";
    ELdkFiles["trusted_peer_node_ids"] = "trusted_peer_node_ids.json";
    ELdkFiles["unconfirmed_transactions"] = "unconfirmed_transactions.json";
    ELdkFiles["broadcasted_transactions"] = "broadcasted_transactions.json";
    ELdkFiles["confirmed_broadcasted_transactions"] = "confirmed_broadcasted_transactions.json";
    ELdkFiles["payment_ids"] = "payment_ids.json";
    ELdkFiles["spendable_outputs"] = "spendable_outputs.json";
    ELdkFiles["payments_claimed"] = "payments_claimed.json";
    ELdkFiles["payments_sent"] = "payments_sent.json";
    ELdkFiles["bolt11_invoices"] = "bolt11_invoices.json";
    ELdkFiles["addresses"] = "addresses.json";
    ELdkFiles["confirmed_watch_outputs"] = "confirmed_watch_outputs.json";
})(ELdkFiles || (ELdkFiles = {}));
export var ELdkData;
(function (ELdkData) {
    ELdkData["channel_manager"] = "channel_manager";
    ELdkData["channel_monitors"] = "channel_monitors";
    ELdkData["peers"] = "peers";
    ELdkData["unconfirmed_transactions"] = "unconfirmed_transactions";
    ELdkData["broadcasted_transactions"] = "broadcasted_transactions";
    ELdkData["payment_ids"] = "payment_ids";
    ELdkData["timestamp"] = "timestamp";
    ELdkData["spendable_outputs"] = "spendable_outputs";
    ELdkData["payments_claimed"] = "payments_claimed";
    ELdkData["payments_sent"] = "payments_sent";
    ELdkData["bolt11_invoices"] = "bolt11_invoices";
})(ELdkData || (ELdkData = {}));
